{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9664aa2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import re\n",
    "\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "import gdt\n",
    "\n",
    "RE_ID = re.compile(r\"ID=([^;]+)\")\n",
    "RE_NAME = re.compile(r\"Name=([^;]+)\")\n",
    "RE_PRODUCT = re.compile(r\"product=([^;]+)\")\n",
    "RE_DESCRIPTION = re.compile(r\"description=([^;]+)\")\n",
    "RE_PARENT = re.compile(r\"Parent=([^;]+)\")\n",
    "RE_GENE = re.compile(r\"gene=([^;]+)\")\n",
    "RE_GENE_SYNONYM = re.compile(r\"gene_synonym=([^;]+)\")\n",
    "RE_NOTE = re.compile(r\"Note=([^;]+)\")\n",
    "\n",
    "\n",
    "def increment_gdt_file(path: Path) -> tuple[Path, int]:\n",
    "    \"\"\"\n",
    "    Increment the GDT file name by 1.\n",
    "    Example: fungi-ncbi_pilot_03.gdt -> fungi-ncbi_pilot_04.gdt\n",
    "    \"\"\"\n",
    "    plist = path.stem.split(\"_\")\n",
    "    try:\n",
    "        number = int(plist[-1]) + 1\n",
    "        plist[-1] = f\"{number:02d}\"\n",
    "    except ValueError:\n",
    "        raise ValueError(\n",
    "            f\"Invalid GDT file name: {path.name}. Expected format: <preferred_name>_##.gdt, where ## is a number.\"\n",
    "        )\n",
    "    return path.parent / f'{\"_\".join(plist)}{path.suffix}', number\n",
    "\n",
    "\n",
    "def get_most_recent_gdt(dir_path: Path, prefix=\"TEMP_\") -> Path:\n",
    "    \"\"\"\n",
    "    Get the most recent gdt file in the directory.\n",
    "    Arguments:\n",
    "        dir_path (Path): Directory to search for GDT files.\n",
    "        prefix (str): Prefix of the GDT files to search for. It will match files like \"<prefix>*.gdt\".\n",
    "    Returns:\n",
    "        Path: Path to the most recent GDT file.\n",
    "    \"\"\"\n",
    "    temp_files = list(dir_path.glob(f\"{prefix}*.gdt\"))\n",
    "    if not temp_files:\n",
    "        return dir_path / f\"{prefix}00.gdt\"\n",
    "    return gdt.gene_dict_impl.natural_sort(temp_files, key=lambda x: x.stem)[-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cc6ac25",
   "metadata": {},
   "source": [
    "\\# TODO  \n",
    "Rather than manually tracking iterations with an 'nth_iteration' variable as implemented in 'AN_missing_gene_dict',  \n",
    "this script employs an automated versioning approach. The system expects the most recent GDT file to follow  \n",
    "the naming convention <preferred_name>\\_pilot\\_##.gdt (e.g., fungi-ncbi_pilot_03.gdt). When modifications are made  \n",
    "to the gene dictionary, the program automatically increments the two-digit suffix to generate the next  \n",
    "version (e.g., analysis_pilot_04.gdt), ensuring seamless version control without manual intervention."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0fff3a64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chosen feature query string: 'type == ['gene', 'tRNA', 'rRNA']'\n"
     ]
    }
   ],
   "source": [
    "# Defines all the global variables used in the script.\n",
    "# Change these variables to match your local setup.\n",
    "# The most_recent_gdt_file variable should be set to the path of the most recent GDT file,\n",
    "# OR the stripped GDT file used in filter command, if applicable.\n",
    "\n",
    "DATA_DIR = \"../sandbox/fungi_mt_model\"\n",
    "most_recent_gdt_file = \"fungi_mt_model_pilot_03.gdt\"\n",
    "global_query_string = gdt.gff3_utils.QS_GENE_TRNA_RRNA\n",
    "remove_orfs = True\n",
    "organelle_type = \"MT\"\n",
    "gff_suffix = \".gff3\"\n",
    "\n",
    "print(f\"Chosen feature query string: '{global_query_string}'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b37b27fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if all variables exist\n",
    "DATA_DIR = Path(DATA_DIR).resolve()\n",
    "MISC_DIR = DATA_DIR / \"misc\"\n",
    "MISC_DIR.mkdir(exist_ok=True)\n",
    "GDT_dir = MISC_DIR / \"gdt\"\n",
    "GDT_dir.mkdir(exist_ok=True)\n",
    "AN_missing_dbxref = MISC_DIR / \"AN_missing_dbxref.txt\"\n",
    "\n",
    "if not DATA_DIR.exists() and not DATA_DIR.is_dir():\n",
    "    raise FileNotFoundError(\n",
    "        f\"Data directory {DATA_DIR} does not exist or is not a directory.\"\n",
    "    )\n",
    "\n",
    "if not AN_missing_dbxref.exists() and not AN_missing_dbxref.is_file():\n",
    "    raise FileNotFoundError(\n",
    "        f\"AN missing dbxref {AN_missing_dbxref} does not exist or is not a file.\"\n",
    "    )\n",
    "\n",
    "most_recent_gdt_file = GDT_dir / most_recent_gdt_file\n",
    "\n",
    "if not most_recent_gdt_file:\n",
    "    print(\n",
    "        \"If you set up a stripped GDT file, please set the path to it in the most_recent_gdt_file variable.\"\n",
    "    )\n",
    "    print(\"Otherwise, ignore this message.\")\n",
    "else:\n",
    "    most_recent_gdt_file = Path(most_recent_gdt_file).resolve()\n",
    "    if not most_recent_gdt_file.exists() and not most_recent_gdt_file.is_file():\n",
    "        raise FileNotFoundError(\n",
    "            f\"Most recent GDT file {most_recent_gdt_file} does not exist or is not a file.\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ddbed68",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-06-06 18:21:24,072 - DEBUG - Logger setup complete. Logging to /home/brenodupin/matheus/gdt/sandbox/fungi_mt_model/misc/02_missing_dbxref.log\n",
      "2025-06-06 18:21:24,079 - DEBUG - Running from notebook AN_missing_dbxref\n"
     ]
    }
   ],
   "source": [
    "_, log = gdt.logger_setup.logger_creater(\n",
    "    log_file=MISC_DIR / \"02_missing_dbxref.log\", console_level=\"DEBUG\", file_level=\"TRACE\"\n",
    ")\n",
    "log.debug(\"Running from notebook AN_missing_dbxref\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "15a53d8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(ANs): 2\n"
     ]
    }
   ],
   "source": [
    "with open(AN_missing_dbxref, \"r\") as f:\n",
    "    ANs = [line.strip() for line in f.readlines() if line.strip()]\n",
    "print(f\"len(ANs): {len(ANs)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f0f3d2d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded gene_dict from /home/brenodupin/matheus/gdt/sandbox/fungi_mt_model/misc/gdt/fungi_mt_model_pilot_03.gdt\n",
      "Header:\n",
      "version 0.0.2\n",
      "Fungi_mt\n",
      "2025-04-09 17:54 - Conversion from gdt to gdt2\n",
      "2025-06-05 18:11 - Stripped GDT version from original GDT file Fungi_mt.gdt\n",
      "2025-06-06 18:07 - Data added from TEMP 01\n",
      "2025-06-06 18:08 - Data added from TEMP Symbol 1\n",
      "automated insertion from missing_dbxref_feature_name\n",
      "\n",
      "GDT Info:\n",
      "Gene dictionary length: 862\n",
      "Label: 56\n",
      "GeneDescription: 535\n",
      "GeneGenerics: 93\n",
      "GeneDbxref: 234\n"
     ]
    }
   ],
   "source": [
    "# Load the GDT file\n",
    "if most_recent_gdt_file:\n",
    "    gene_dict = gdt.gene_dict_impl.create_gene_dict(\n",
    "        most_recent_gdt_file, max_an_sources=0\n",
    "    )\n",
    "    print(f\"Loaded gene_dict from {most_recent_gdt_file}\\nHeader:\")\n",
    "    [print(x) for x in gene_dict[\"gdt_header\"]]\n",
    "    print(\"\\nGDT Info:\")\n",
    "    [print(x) for x in gene_dict[\"gdt_info\"]]\n",
    "else:\n",
    "    gene_dict = {}\n",
    "    print(\"No GDT file found, starting with an empty gene_dict.\")\n",
    "\n",
    "temp_gene_dict = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d985e87",
   "metadata": {},
   "source": [
    "### Deeper investigation using other gff attributes, primarily gene="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c3d1c41",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-06-06 18:21:39,106 - DEBUG - missing_dbxref: creation of features_info_df\n"
     ]
    }
   ],
   "source": [
    "temp_list = []\n",
    "log.debug(\"missing_dbxref: creation of features_info_df\")\n",
    "for AN in ANs:\n",
    "    an_path = DATA_DIR / f\"{AN}{gff_suffix}\"\n",
    "    df = gdt.gff3_utils.load_gff3(\n",
    "        an_path, query_string=global_query_string, usecols=gdt.GFF3_COLUMNS\n",
    "    )\n",
    "    df = gdt.gff3_utils.filter_orfs(df) if remove_orfs else df\n",
    "\n",
    "    df[\"gene_id\"] = df[\"attributes\"].str.extract(RE_ID, expand=False)\n",
    "    df = df[~df[\"gene_id\"].isin(gene_dict)]\n",
    "\n",
    "    df[\"name\"] = df[\"attributes\"].str.extract(RE_NAME, expand=False)\n",
    "    df[\"product\"] = df[\"attributes\"].str.extract(RE_PRODUCT, expand=False)\n",
    "    df[\"description\"] = df[\"attributes\"].str.extract(RE_DESCRIPTION, expand=False)\n",
    "    df[\"gene\"] = df[\"attributes\"].str.extract(RE_GENE, expand=False)\n",
    "    df[\"parent\"] = df[\"attributes\"].str.extract(RE_PARENT, expand=False)\n",
    "    df[\"gene_synonym\"] = df[\"attributes\"].str.extract(RE_GENE_SYNONYM, expand=False)\n",
    "    df[\"note\"] = df[\"attributes\"].str.extract(RE_NOTE, expand=False)\n",
    "\n",
    "    if (\n",
    "        df[[\"name\", \"product\", \"description\", \"gene\", \"gene_synonym\", \"note\"]]\n",
    "        .isna()\n",
    "        .all(axis=1)\n",
    "        .any()\n",
    "    ):\n",
    "        log.debug(f\"Warning: {AN} has row(s) with no identifiable atribute.\")\n",
    "        log.debug(\n",
    "            \"Please modify this script to add a new possible identifiable attribute or just remove the AN from the list.\"\n",
    "        )\n",
    "        log.debug(\n",
    "            df[\n",
    "                df[[\"name\", \"product\", \"description\", \"gene\", \"gene_synonym\", \"note\"]]\n",
    "                .isna()\n",
    "                .all(axis=1)\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    temp_list.extend(df.to_dict(\"records\"))\n",
    "\n",
    "features_info_df = pd.DataFrame(temp_list)\n",
    "features_info_df = features_info_df.drop(\n",
    "    columns=[\"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"phase\", \"attributes\"]\n",
    ")\n",
    "\n",
    "dc = [\n",
    "    col\n",
    "    for col in [\"name\", \"product\", \"description\", \"gene\", \"gene_synonym\", \"note\"]\n",
    "    if features_info_df[col].isna().all()\n",
    "]\n",
    "\n",
    "features_info_df[\"feature_name\"] = (\n",
    "    features_info_df[\"gene\"]\n",
    "    .fillna(features_info_df[\"product\"])\n",
    "    .fillna(features_info_df[\"description\"])\n",
    "    .fillna(features_info_df[\"name\"])\n",
    "    .fillna(features_info_df[\"note\"])\n",
    "    .fillna(features_info_df[\"gene_synonym\"])\n",
    ")\n",
    "\n",
    "features_info_df = features_info_df.drop(columns=dc)\n",
    "features_info_df = features_info_df.sort_values(by=\"feature_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4c3d1c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_gdt_compliance = True\n",
    "comment = \"Manual from missing_dbxref_names_raw\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ff3ed357",
   "metadata": {},
   "outputs": [],
   "source": [
    "if add_gdt_compliance:\n",
    "    gdt_str = f' #gd MANUAL{ \" #c \" + comment if comment else \"\" }'\n",
    "else:\n",
    "    gdt_str = \"\"\n",
    "\n",
    "# df with 2 columns, one for feature_names and one for in_gene_dict\n",
    "new_df = pd.DataFrame({\"feature_name\": features_info_df[\"feature_name\"].unique()})\n",
    "new_df[\"in_gene_dict\"] = new_df[\"feature_name\"].isin(gene_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84cd8111",
   "metadata": {},
   "source": [
    "#### Check feature name, where both df and gene_dict use .lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "10ec68d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "check_df = new_df[~new_df[\"in_gene_dict\"]].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ce88c095",
   "metadata": {},
   "outputs": [],
   "source": [
    "check_df[\"lower_feature_name\"] = check_df[\"feature_name\"].str.lower()\n",
    "check_df[\"in_gene_dict_lower\"] = check_df[\"lower_feature_name\"].isin(gene_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1fb7e319",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_name</th>\n",
       "      <th>in_gene_dict</th>\n",
       "      <th>lower_feature_name</th>\n",
       "      <th>in_gene_dict_lower</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>Ycf4</td>\n",
       "      <td>False</td>\n",
       "      <td>ycf4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    feature_name  in_gene_dict lower_feature_name  in_gene_dict_lower\n",
       "248         Ycf4         False               ycf4                True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_df[check_df[\"in_gene_dict_lower\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "064bff77",
   "metadata": {},
   "outputs": [],
   "source": [
    "check_df[\"in_gene_dict_both_lower\"] = check_df[\"lower_feature_name\"].isin(\n",
    "    [x.lower() for x in gene_dict]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "88c27bd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_name</th>\n",
       "      <th>in_gene_dict</th>\n",
       "      <th>lower_feature_name</th>\n",
       "      <th>in_gene_dict_lower</th>\n",
       "      <th>in_gene_dict_both_lower</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>Ycf4</td>\n",
       "      <td>False</td>\n",
       "      <td>ycf4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>ftsH</td>\n",
       "      <td>False</td>\n",
       "      <td>ftsh</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>rnpB</td>\n",
       "      <td>False</td>\n",
       "      <td>rnpb</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>rps2b</td>\n",
       "      <td>False</td>\n",
       "      <td>rps2b</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    feature_name  in_gene_dict lower_feature_name  in_gene_dict_lower  \\\n",
       "248         Ycf4         False               ycf4                True   \n",
       "382         ftsH         False               ftsh               False   \n",
       "434         rnpB         False               rnpb               False   \n",
       "470        rps2b         False              rps2b               False   \n",
       "\n",
       "     in_gene_dict_both_lower  \n",
       "248                     True  \n",
       "382                     True  \n",
       "434                     True  \n",
       "470                     True  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_df[check_df[\"in_gene_dict_both_lower\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5320d919",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_names = new_df[~new_df[\"in_gene_dict\"]][\"feature_name\"]\n",
    "unique_names = gdt.gene_dict_impl.natural_sort(unique_names)\n",
    "with open(MISC_DIR / \"feature_name.txt\", \"w+\") as f1:\n",
    "    for name in unique_names:\n",
    "        f1.write(f\"{name}{gdt_str}\\n\")\n",
    "\n",
    "features_info_df.to_csv(MISC_DIR / \"features_info.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d13888a",
   "metadata": {},
   "source": [
    "The user must now parse feature_names.txt  \n",
    "\n",
    "Features that can be easily identifiable must be added to the current  \n",
    "version of the gdt, and features that need deeper investigation should be  \n",
    "copied to a new file name 'features_unk.txt'\n",
    "  \n",
    "The program will now try to automatically add the gene_ids with feature name  \n",
    "that __is not__ in 'features_unk.txt'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6a6dd89e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if the names exist in the gene_dict\n",
    "features_info_df = pd.read_csv(MISC_DIR / \"features_info.tsv\", sep=\"\\t\")\n",
    "\n",
    "names_unk = set()\n",
    "with open(MISC_DIR / \"features_unk.txt\", \"r\") as f1:\n",
    "    for line in f1:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue\n",
    "\n",
    "        if \"#gd\" in line:\n",
    "            line = line.split(\"#gd\")[0].strip()\n",
    "\n",
    "        names_unk.add(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef05f9d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_not_in_dict = set()\n",
    "all_names = set(features_info_df[\"feature_name\"].unique()) - names_unk\n",
    "for name in all_names:\n",
    "    if name not in gene_dict:\n",
    "        names_not_in_dict.add(name)\n",
    "\n",
    "if names_not_in_dict:\n",
    "    log.debug(f\"Warning: {len(names_not_in_dict)} names not in gene_dict!\")\n",
    "    log.debug(\n",
    "        \"These names are not in feature_unk, so you marked them as identifiable. Please identify them or add them feature_unk.\"\n",
    "    )\n",
    "    log.debug(\n",
    "        \"It could also be that you forgot to reload the gene_dict with the changes that you made.\"\n",
    "    )\n",
    "    [log.debug(name) for name in names_not_in_dict]\n",
    "    raise ValueError(f\"Error: {len(names_not_in_dict)} names not in gene_dict!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5ad26831",
   "metadata": {},
   "outputs": [],
   "source": [
    "comment = \"automated insertion from missing_dbxref_feature_name\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "80101ad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_gene_ids = features_info_df[~features_info_df[\"feature_name\"].isin(names_unk)][\n",
    "    \"gene_id\"\n",
    "].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "24db6689",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['gene-atp6', 'gene-atp8', 'gene-atp9', 'gene-cob', 'gene-cox1',\n",
       "       'gene-cox2', 'gene-cox3', 'rna-AFUA_m0250', 'rna-lsu', 'gene-lsu',\n",
       "       'gene-rpm1', 'rna-rpm1', 'rna-AFUA_m0120', 'rna-ssu', 'gene-ssu',\n",
       "       'rna-AFUA_m0330', 'gene-tRNA-Ala', 'rna-AFUA_m0470',\n",
       "       'rna-tRNA-Ala', 'gene-tRNA-Arg', 'rna-tRNA-Arg', 'rna-AFUA_m0500',\n",
       "       'gene-tRNA-Arg-2', 'rna-AFUA_m0070', 'rna-tRNA-Arg-2',\n",
       "       'gene-tRNA-Asn', 'rna-tRNA-Asn', 'rna-AFUA_m0080',\n",
       "       'rna-AFUA_m0030', 'gene-tRNA-Asp', 'rna-AFUA_m0190',\n",
       "       'rna-tRNA-Asp', 'rna-AFUA_m0020', 'gene-tRNA-Cys', 'rna-tRNA-Cys',\n",
       "       'gene-tRNA-Gln', 'rna-AFUA_m0360', 'rna-tRNA-Gln', 'gene-tRNA-Glu',\n",
       "       'rna-AFUA_m0280', 'rna-tRNA-Glu', 'rna-AFUA_m0180',\n",
       "       'rna-AFUA_m0170', 'gene-tRNA-Gly', 'rna-tRNA-Gly', 'rna-tRNA-His',\n",
       "       'rna-AFUA_m0410', 'gene-tRNA-His', 'rna-AFUA_m0050',\n",
       "       'rna-AFUA_m0380', 'gene-tRNA-Ile', 'rna-AFUA_m0100',\n",
       "       'rna-tRNA-Ile', 'rna-AFUA_m0220', 'rna-tRNA-Leu-2',\n",
       "       'gene-tRNA-Leu', 'rna-tRNA-Leu', 'rna-AFUA_m0350',\n",
       "       'rna-AFUA_m0320', 'gene-tRNA-Leu-2', 'rna-AFUA_m0160',\n",
       "       'gene-tRNA-Lys', 'rna-tRNA-Lys', 'rna-AFUA_m0370', 'rna-tRNA-Met',\n",
       "       'rna-AFUA_m0310', 'gene-tRNA-Met-2', 'rna-tRNA-Met-2',\n",
       "       'rna-AFUA_m0300', 'gene-tRNA-Met', 'rna-AFUA_m0340',\n",
       "       'rna-tRNA-Phe', 'gene-tRNA-Phe', 'rna-AFUA_m0240', 'rna-tRNA-Pro',\n",
       "       'gene-tRNA-Pro', 'rna-AFUA_m0210', 'gene-tRNA-Ser-2',\n",
       "       'rna-tRNA-Ser', 'rna-AFUA_m0230', 'rna-tRNA-Ser-2',\n",
       "       'gene-tRNA-Ser', 'rna-AFUA_m0200', 'rna-AFUA_m0270',\n",
       "       'rna-tRNA-Trp', 'gene-tRNA-Trp', 'gene-tRNA-Tyr', 'rna-AFUA_m0130',\n",
       "       'rna-tRNA-Tyr', 'rna-tRNA-Val', 'gene-tRNA-Val', 'rna-AFUA_m0290',\n",
       "       'gene-var1'], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e914f421",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_id: gene-atp6, label: MT-ATP6, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-atp8, label: MT-ATP8, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-atp9, label: MT-ATP9, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-cob, label: MT-COB, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-cox1, label: MT-CO1, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-cox2, label: MT-CO2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-cox3, label: MT-CO3, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0250, label: MT-RNR2, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-lsu, label: MT-RNR2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-lsu, label: MT-RNR2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-rpm1, label: MT-RPM1, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-rpm1, label: MT-RPM1, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0120, label: MT-RNR1, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-ssu, label: MT-RNR1, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-ssu, label: MT-RNR1, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0330, label: MT-TA, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Ala, label: MT-TA, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0470, label: MT-TA, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Ala, label: MT-TA, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Arg, label: MT-TR, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Arg, label: MT-TR, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0500, label: MT-TR, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Arg-2, label: MT-TR, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0070, label: MT-TR, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Arg-2, label: MT-TR, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Asn, label: MT-TN, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Asn, label: MT-TN, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0080, label: MT-TN, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0030, label: MT-TN, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Asp, label: MT-TD, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0190, label: MT-TD, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Asp, label: MT-TD, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0020, label: MT-TC, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Cys, label: MT-TC, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Cys, label: MT-TC, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Gln, label: MT-TQ, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0360, label: MT-TQ, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Gln, label: MT-TQ, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Glu, label: MT-TE, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0280, label: MT-TE, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Glu, label: MT-TE, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0180, label: MT-TG, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0170, label: MT-TG, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Gly, label: MT-TG, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Gly, label: MT-TG, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-His, label: MT-TH, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0410, label: MT-TH, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-His, label: MT-TH, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0050, label: MT-TI, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0380, label: MT-TI, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Ile, label: MT-TI, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0100, label: MT-TI, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Ile, label: MT-TI, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0220, label: MT-TI, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Leu-2, label: MT-TL1|2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Leu, label: MT-TL1|2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Leu, label: MT-TL1|2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0350, label: MT-TL1|2, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0320, label: MT-TL1|2, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Leu-2, label: MT-TL1|2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0160, label: MT-TK, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Lys, label: MT-TK, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Lys, label: MT-TK, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0370, label: MT-TM, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Met, label: MT-TM, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0310, label: MT-TM, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Met-2, label: MT-TM, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Met-2, label: MT-TM, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0300, label: MT-TM, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Met, label: MT-TM, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0340, label: MT-TF, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Phe, label: MT-TF, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Phe, label: MT-TF, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0240, label: MT-TP, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Pro, label: MT-TP, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Pro, label: MT-TP, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0210, label: MT-TSEC, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Ser-2, label: MT-TS1|2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Ser, label: MT-TS1|2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0230, label: MT-TS1|2, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Ser-2, label: MT-TS1|2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Ser, label: MT-TS1|2, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0200, label: MT-TS1|2, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0270, label: MT-TT, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Trp, label: MT-TW, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Trp, label: MT-TW, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Tyr, label: MT-TY, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0130, label: MT-TY, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Tyr, label: MT-TY, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-tRNA-Val, label: MT-TV, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-tRNA-Val, label: MT-TV, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: rna-AFUA_m0290, label: MT-TV, an_sources: ['JQ346808.1'], comment: automated insertion from missing_dbxref_feature_name\n",
      "gene_id: gene-var1, label: MT-VAR1, an_sources: ['HE983613.1'], comment: automated insertion from missing_dbxref_feature_name\n"
     ]
    }
   ],
   "source": [
    "unique_gene_ids = features_info_df[~features_info_df[\"feature_name\"].isin(names_unk)][\n",
    "    \"gene_id\"\n",
    "].unique()\n",
    "\n",
    "for gene_id in unique_gene_ids:\n",
    "    df = features_info_df[features_info_df[\"gene_id\"] == gene_id]\n",
    "\n",
    "    # sanity check, are all feature_names the same?\n",
    "    if df[\"feature_name\"].nunique() != 1:\n",
    "        print(\n",
    "            f\"Warning: {gene_id} has multiple feature_names: {df['feature_name'].unique()}\"\n",
    "        )\n",
    "        print(\"\\tChecking if they have the same label in gene_dict...\")\n",
    "\n",
    "        labels = {gene_dict[feat].label for feat in df[\"feature_name\"].unique()}\n",
    "        if len(labels) != 1:\n",
    "            print(f\"\\tError: {gene_id} has multiple labels: {labels}\")\n",
    "            continue\n",
    "        else:\n",
    "            print(f\"\\tAll feature_names have the same label: {labels.pop()}\")\n",
    "\n",
    "    label = gene_dict[df[\"feature_name\"].iloc[0]].label\n",
    "    an_sources = df[\"seqid\"].unique().tolist()\n",
    "    print(\n",
    "        f\"gene_id: {gene_id}, label: {label}, an_sources: {an_sources}, comment: {comment}\"\n",
    "    )\n",
    "    gene_dict[gene_id] = gdt.gene_dict_impl.GeneGeneric(\n",
    "        label=label, an_sources=an_sources, c=comment\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82329a80",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-06-06 18:26:43,718 - INFO - Writing gene_dict file: /home/brenodupin/matheus/gdt/sandbox/fungi_mt_model/misc/gdt/fungi_mt_model_pilot_03.gdt | Iteration: 3\n"
     ]
    }
   ],
   "source": [
    "if gene_dict:\n",
    "    new_path, nth_iteration = increment_gdt_file(most_recent_gdt_file)\n",
    "    log.info(f\"Writing gene_dict file: {new_path} | Iteration: {nth_iteration}\")\n",
    "    gene_dict[\"gdt_info\"] = gdt.gene_dict_impl.get_gene_dict_info(gene_dict)\n",
    "    gene_dict[\"gdt_header\"].append(\n",
    "        \"automated insertion from missing_dbxref_feature_name\"\n",
    "    )\n",
    "    gdt.gene_dict_impl.write_gdt_file(gene_dict, new_path, overwrite=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "350d5c33",
   "metadata": {},
   "source": [
    "#### missing_dbxref_names_unk.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d9262ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_df_rows(cds_trna):\n",
    "    for row_cds in cds_trna.itertuples():\n",
    "        print(\n",
    "            f\"\\tan: {row_cds.seqid}| gene_id: {row_cds.gene_id}| p: {row_cds.parent}| fn: {row_cds.feature_name}\"\n",
    "        )\n",
    "\n",
    "\n",
    "def print_df_rows_log(df, log):\n",
    "    for r in df.itertuples():\n",
    "        log.debug(\n",
    "            f\"\\tan: {r.seqid}| gene_id: {r.gene_id}| p: {r.parent}| fn: {r.feature_name}\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "fdeb51ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_unk = set()\n",
    "with open(MISC_DIR / \"features_unk.txt\", \"r\") as f1:\n",
    "    for line in f1:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue\n",
    "\n",
    "        if \"#gd\" in line:\n",
    "            line = line.split(\"#gd\")[0].strip()\n",
    "\n",
    "        names_unk.add(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d5eea968",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_info_df = pd.read_csv(MISC_DIR / \"features_info.tsv\", sep=\"\\t\")\n",
    "features_know_df = features_info_df[~features_info_df[\"feature_name\"].isin(names_unk)]\n",
    "features_unk_df = (\n",
    "    features_info_df[features_info_df[\"feature_name\"].isin(names_unk)]\n",
    "    .copy()\n",
    "    .reset_index(drop=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f823a585",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_unk_df[\"feature_child\"] = np.nan\n",
    "# make feature_child a object type\n",
    "features_unk_df[\"feature_child\"] = features_unk_df[\"feature_child\"].astype(object)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3660d66d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-06-03 19:29:20,325 - DEBUG - missing_dbxref: matching probable 'child feature + parent gene' pair (1st pass, matching rows of known features of features_info.tsv)\n",
      "2025-06-03 19:29:20,330 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0010\n",
      "2025-06-03 19:29:20,331 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0020\n",
      "2025-06-03 19:29:20,333 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0030\n",
      "2025-06-03 19:29:20,334 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0040\n",
      "2025-06-03 19:29:20,335 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0050\n",
      "2025-06-03 19:29:20,336 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0060\n",
      "2025-06-03 19:29:20,337 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0070\n",
      "2025-06-03 19:29:20,339 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0080\n",
      "2025-06-03 19:29:20,340 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0090\n",
      "2025-06-03 19:29:20,341 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0100\n",
      "2025-06-03 19:29:20,344 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0110\n",
      "2025-06-03 19:29:20,347 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0120\n",
      "2025-06-03 19:29:20,348 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0130\n",
      "2025-06-03 19:29:20,349 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0140\n",
      "2025-06-03 19:29:20,351 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0150\n",
      "2025-06-03 19:29:20,353 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0160\n",
      "2025-06-03 19:29:20,355 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0170\n",
      "2025-06-03 19:29:20,357 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0180\n",
      "2025-06-03 19:29:20,358 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0190\n",
      "2025-06-03 19:29:20,359 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0200\n",
      "2025-06-03 19:29:20,360 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0210\n",
      "2025-06-03 19:29:20,363 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0220\n",
      "2025-06-03 19:29:20,365 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0230\n",
      "2025-06-03 19:29:20,366 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0240\n",
      "2025-06-03 19:29:20,367 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0250\n",
      "2025-06-03 19:29:20,369 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0260\n",
      "2025-06-03 19:29:20,370 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0270\n",
      "2025-06-03 19:29:20,372 - DEBUG - Found perfect candidate for gene-COCOBI_pt-0280: tRNA-Asp\n",
      "2025-06-03 19:29:20,374 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0290\n",
      "2025-06-03 19:29:20,375 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0300\n",
      "2025-06-03 19:29:20,378 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0310\n",
      "2025-06-03 19:29:20,379 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0320\n",
      "2025-06-03 19:29:20,382 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0330\n",
      "2025-06-03 19:29:20,384 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0340\n",
      "2025-06-03 19:29:20,386 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0350\n",
      "2025-06-03 19:29:20,387 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0360\n",
      "2025-06-03 19:29:20,389 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0370\n",
      "2025-06-03 19:29:20,392 - DEBUG - Found perfect candidate for gene-COCOBI_pt-0380: tRNA-Met\n",
      "2025-06-03 19:29:20,394 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0390\n",
      "2025-06-03 19:29:20,396 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0400\n",
      "2025-06-03 19:29:20,397 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0410\n",
      "2025-06-03 19:29:20,399 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0420\n",
      "2025-06-03 19:29:20,401 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0430\n",
      "2025-06-03 19:29:20,402 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0440\n",
      "2025-06-03 19:29:20,404 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0450\n",
      "2025-06-03 19:29:20,405 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0460\n",
      "2025-06-03 19:29:20,406 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0470\n",
      "2025-06-03 19:29:20,407 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0480\n",
      "2025-06-03 19:29:20,409 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0490\n",
      "2025-06-03 19:29:20,411 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0500\n",
      "2025-06-03 19:29:20,413 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0510\n",
      "2025-06-03 19:29:20,414 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0520\n",
      "2025-06-03 19:29:20,417 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0530\n",
      "2025-06-03 19:29:20,418 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0540\n",
      "2025-06-03 19:29:20,419 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0550\n",
      "2025-06-03 19:29:20,423 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0560\n",
      "2025-06-03 19:29:20,425 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0570\n",
      "2025-06-03 19:29:20,428 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0580\n",
      "2025-06-03 19:29:20,429 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0590\n",
      "2025-06-03 19:29:20,430 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0600\n",
      "2025-06-03 19:29:20,433 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0610\n",
      "2025-06-03 19:29:20,434 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0620\n",
      "2025-06-03 19:29:20,436 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0630\n",
      "2025-06-03 19:29:20,438 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0640\n",
      "2025-06-03 19:29:20,439 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0650\n",
      "2025-06-03 19:29:20,441 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0660\n",
      "2025-06-03 19:29:20,446 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0670\n",
      "2025-06-03 19:29:20,449 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0680\n",
      "2025-06-03 19:29:20,451 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0690\n",
      "2025-06-03 19:29:20,452 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0700\n",
      "2025-06-03 19:29:20,454 - DEBUG - Found perfect candidate for gene-COCOBI_pt-0710: tRNA-Phe\n",
      "2025-06-03 19:29:20,456 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0720\n",
      "2025-06-03 19:29:20,457 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0730\n",
      "2025-06-03 19:29:20,458 - DEBUG - Found perfect candidate for gene-COCOBI_pt-0740: tRNA-Thr\n",
      "2025-06-03 19:29:20,459 - DEBUG - Found perfect candidate for gene-COCOBI_pt-0750: tRNA-Gln\n",
      "2025-06-03 19:29:20,461 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0760\n",
      "2025-06-03 19:29:20,462 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0770\n",
      "2025-06-03 19:29:20,463 - DEBUG - Found perfect candidate for gene-COCOBI_pt-0780: tRNA-Lys\n",
      "2025-06-03 19:29:20,466 - DEBUG - Found perfect candidate for gene-COCOBI_pt-0790: tRNA-His\n",
      "2025-06-03 19:29:20,470 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0800\n",
      "2025-06-03 19:29:20,472 - DEBUG - Found perfect candidate for gene-COCOBI_pt-0810: tRNA-Gly\n",
      "2025-06-03 19:29:20,474 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0820\n",
      "2025-06-03 19:29:20,475 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0830\n",
      "2025-06-03 19:29:20,478 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0840\n",
      "2025-06-03 19:29:20,479 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0850\n",
      "2025-06-03 19:29:20,481 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0860\n",
      "2025-06-03 19:29:20,483 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0870\n",
      "2025-06-03 19:29:20,484 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0880\n",
      "2025-06-03 19:29:20,487 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0890\n",
      "2025-06-03 19:29:20,488 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0900\n",
      "2025-06-03 19:29:20,492 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0910\n",
      "2025-06-03 19:29:20,495 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0920\n",
      "2025-06-03 19:29:20,496 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0930\n",
      "2025-06-03 19:29:20,497 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0940\n",
      "2025-06-03 19:29:20,500 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0950\n",
      "2025-06-03 19:29:20,501 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0960\n",
      "2025-06-03 19:29:20,503 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0970\n",
      "2025-06-03 19:29:20,505 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0980\n",
      "2025-06-03 19:29:20,506 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-0990\n",
      "2025-06-03 19:29:20,508 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1000\n",
      "2025-06-03 19:29:20,509 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1010: 23S ribosomal RNA\n",
      "2025-06-03 19:29:20,512 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1020: tRNA-Ala\n",
      "2025-06-03 19:29:20,518 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1030: tRNA-Ile\n",
      "2025-06-03 19:29:20,522 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1040\n",
      "2025-06-03 19:29:20,524 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1050\n",
      "2025-06-03 19:29:20,525 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1060\n",
      "2025-06-03 19:29:20,526 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1070\n",
      "2025-06-03 19:29:20,528 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1080\n",
      "2025-06-03 19:29:20,529 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1090\n",
      "2025-06-03 19:29:20,530 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1100\n",
      "2025-06-03 19:29:20,531 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1110\n",
      "2025-06-03 19:29:20,533 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1120: tRNA-Met\n",
      "2025-06-03 19:29:20,534 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1130\n",
      "2025-06-03 19:29:20,536 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1140\n",
      "2025-06-03 19:29:20,538 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1150\n",
      "2025-06-03 19:29:20,540 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1160: tRNA-Leu\n",
      "2025-06-03 19:29:20,542 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1170\n",
      "2025-06-03 19:29:20,545 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1180\n",
      "2025-06-03 19:29:20,546 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1190\n",
      "2025-06-03 19:29:20,548 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1200\n",
      "2025-06-03 19:29:20,549 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1210\n",
      "2025-06-03 19:29:20,550 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1220\n",
      "2025-06-03 19:29:20,551 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1230\n",
      "2025-06-03 19:29:20,553 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1240\n",
      "2025-06-03 19:29:20,554 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1250: tRNA-Thr\n",
      "2025-06-03 19:29:20,556 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1260\n",
      "2025-06-03 19:29:20,558 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1270\n",
      "2025-06-03 19:29:20,561 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1280\n",
      "2025-06-03 19:29:20,563 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1290\n",
      "2025-06-03 19:29:20,565 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1300\n",
      "2025-06-03 19:29:20,566 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1310\n",
      "2025-06-03 19:29:20,568 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1320\n",
      "2025-06-03 19:29:20,573 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1330\n",
      "2025-06-03 19:29:20,575 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1340\n",
      "2025-06-03 19:29:20,577 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1350\n",
      "2025-06-03 19:29:20,578 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1360\n",
      "2025-06-03 19:29:20,580 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1370: tRNA-Trp\n",
      "2025-06-03 19:29:20,582 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1380: tRNA-Pro\n",
      "2025-06-03 19:29:20,583 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1390\n",
      "2025-06-03 19:29:20,584 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1400\n",
      "2025-06-03 19:29:20,587 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1410: tRNA-Arg\n",
      "2025-06-03 19:29:20,588 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1420\n",
      "2025-06-03 19:29:20,590 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1430\n",
      "2025-06-03 19:29:20,594 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1440\n",
      "2025-06-03 19:29:20,596 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1450\n",
      "2025-06-03 19:29:20,601 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1460\n",
      "2025-06-03 19:29:20,604 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1470\n",
      "2025-06-03 19:29:20,606 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1480\n",
      "2025-06-03 19:29:20,609 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1490\n",
      "2025-06-03 19:29:20,610 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1500: 16S ribosomal RNA\n",
      "2025-06-03 19:29:20,614 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1510\n",
      "2025-06-03 19:29:20,616 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1520\n",
      "2025-06-03 19:29:20,617 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1530\n",
      "2025-06-03 19:29:20,620 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1540\n",
      "2025-06-03 19:29:20,622 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1550\n",
      "2025-06-03 19:29:20,626 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1560\n",
      "2025-06-03 19:29:20,628 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1570\n",
      "2025-06-03 19:29:20,630 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1580\n",
      "2025-06-03 19:29:20,631 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1590\n",
      "2025-06-03 19:29:20,632 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1600\n",
      "2025-06-03 19:29:20,634 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1610: tRNA-Cys\n",
      "2025-06-03 19:29:20,636 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1620\n",
      "2025-06-03 19:29:20,638 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1630\n",
      "2025-06-03 19:29:20,643 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1640\n",
      "2025-06-03 19:29:20,645 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1650\n",
      "2025-06-03 19:29:20,647 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1660\n",
      "2025-06-03 19:29:20,649 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1670\n",
      "2025-06-03 19:29:20,650 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1680\n",
      "2025-06-03 19:29:20,658 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1690: tRNA-Arg\n",
      "2025-06-03 19:29:20,661 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1700\n",
      "2025-06-03 19:29:20,662 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1710\n",
      "2025-06-03 19:29:20,664 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1720: tRNA-Leu\n",
      "2025-06-03 19:29:20,666 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1730: tRNA-Tyr\n",
      "2025-06-03 19:29:20,667 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1740\n",
      "2025-06-03 19:29:20,669 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1750\n",
      "2025-06-03 19:29:20,673 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1760: tRNA-Arg\n",
      "2025-06-03 19:29:20,675 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1770\n",
      "2025-06-03 19:29:20,677 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1780\n",
      "2025-06-03 19:29:20,679 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1790\n",
      "2025-06-03 19:29:20,681 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1800\n",
      "2025-06-03 19:29:20,682 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1810\n",
      "2025-06-03 19:29:20,684 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1820\n",
      "2025-06-03 19:29:20,687 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1830\n",
      "2025-06-03 19:29:20,689 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1840\n",
      "2025-06-03 19:29:20,690 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1850\n",
      "2025-06-03 19:29:20,693 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1860\n",
      "2025-06-03 19:29:20,694 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1870\n",
      "2025-06-03 19:29:20,695 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1880: tRNA-Gly\n",
      "2025-06-03 19:29:20,697 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1890: tRNA-Ser\n",
      "2025-06-03 19:29:20,699 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1900: tRNA-Val\n",
      "2025-06-03 19:29:20,701 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1910\n",
      "2025-06-03 19:29:20,703 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1920\n",
      "2025-06-03 19:29:20,705 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1930\n",
      "2025-06-03 19:29:20,706 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1940\n",
      "2025-06-03 19:29:20,707 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1950\n",
      "2025-06-03 19:29:20,708 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1960\n",
      "2025-06-03 19:29:20,710 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1970\n",
      "2025-06-03 19:29:20,711 - DEBUG - Found perfect candidate for gene-COCOBI_pt-1980: tRNA-Leu\n",
      "2025-06-03 19:29:20,713 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-1990\n",
      "2025-06-03 19:29:20,717 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2000\n",
      "2025-06-03 19:29:20,719 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2010\n",
      "2025-06-03 19:29:20,721 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2020\n",
      "2025-06-03 19:29:20,723 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2030\n",
      "2025-06-03 19:29:20,726 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2040\n",
      "2025-06-03 19:29:20,728 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2050\n",
      "2025-06-03 19:29:20,730 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2060\n",
      "2025-06-03 19:29:20,732 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2070\n",
      "2025-06-03 19:29:20,733 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2080\n",
      "2025-06-03 19:29:20,735 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2090\n",
      "2025-06-03 19:29:20,737 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2100\n",
      "2025-06-03 19:29:20,738 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2110\n",
      "2025-06-03 19:29:20,740 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2120\n",
      "2025-06-03 19:29:20,743 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2130\n",
      "2025-06-03 19:29:20,747 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2140\n",
      "2025-06-03 19:29:20,748 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2150\n",
      "2025-06-03 19:29:20,750 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2160\n",
      "2025-06-03 19:29:20,751 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2170\n",
      "2025-06-03 19:29:20,753 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2180\n",
      "2025-06-03 19:29:20,755 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2190\n",
      "2025-06-03 19:29:20,757 - DEBUG - Found perfect candidate for gene-COCOBI_pt-2200: tRNA-Arg\n",
      "2025-06-03 19:29:20,759 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2210\n",
      "2025-06-03 19:29:20,761 - DEBUG - Found perfect candidate for gene-COCOBI_pt-2220: tRNA-Ser\n",
      "2025-06-03 19:29:20,763 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2230\n",
      "2025-06-03 19:29:20,764 - DEBUG - Found perfect candidate for gene-COCOBI_pt-2240: tRNA-Leu\n",
      "2025-06-03 19:29:20,765 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2250\n",
      "2025-06-03 19:29:20,767 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2260\n",
      "2025-06-03 19:29:20,769 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2270\n",
      "2025-06-03 19:29:20,770 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2280\n",
      "2025-06-03 19:29:20,771 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2290\n",
      "2025-06-03 19:29:20,772 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2300\n",
      "2025-06-03 19:29:20,774 - DEBUG - Found perfect candidate for gene-COCOBI_pt-2310: tRNA-Asn\n",
      "2025-06-03 19:29:20,775 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2320\n",
      "2025-06-03 19:29:20,779 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2330\n",
      "2025-06-03 19:29:20,781 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2340\n",
      "2025-06-03 19:29:20,783 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2350\n",
      "2025-06-03 19:29:20,785 - DEBUG - Found perfect candidate for gene-COCOBI_pt-2360: tRNA-Met\n",
      "2025-06-03 19:29:20,788 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2370\n",
      "2025-06-03 19:29:20,794 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2380\n",
      "2025-06-03 19:29:20,796 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2390\n",
      "2025-06-03 19:29:20,800 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2400\n",
      "2025-06-03 19:29:20,802 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2410\n",
      "2025-06-03 19:29:20,804 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2420\n",
      "2025-06-03 19:29:20,805 - WARNING - no feature* found in AP025008.1 that have parent equals to gene-COCOBI_pt-2430\n",
      "2025-06-03 19:29:20,807 - WARNING - no feature* found in AP014542.1 that have parent equals to gene-I-CreI\n",
      "2025-06-03 19:29:20,808 - WARNING - no feature* found in AP012291.1 that have parent equals to gene-I-CreI\n",
      "2025-06-03 19:29:20,810 - WARNING - no feature* found in AP018038.1 that have parent equals to gene-Rsub_pt095\n",
      "2025-06-03 19:29:20,811 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00001\n",
      "2025-06-03 19:29:20,814 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00002\n",
      "2025-06-03 19:29:20,815 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00003\n",
      "2025-06-03 19:29:20,816 - DEBUG - Found perfect candidate for gene-cvarcp_00004: tRNA-Tyr\n",
      "2025-06-03 19:29:20,818 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00005\n",
      "2025-06-03 19:29:20,819 - DEBUG - Found perfect candidate for gene-cvarcp_00006: tRNA-Ser\n",
      "2025-06-03 19:29:20,820 - DEBUG - Found perfect candidate for gene-cvarcp_00008: tRNA-Met\n",
      "2025-06-03 19:29:20,823 - DEBUG - Found perfect candidate for gene-cvarcp_00009: tRNA-Glu\n",
      "2025-06-03 19:29:20,824 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00010\n",
      "2025-06-03 19:29:20,825 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00011\n",
      "2025-06-03 19:29:20,827 - DEBUG - Found perfect candidate for gene-cvarcp_00012: tRNA-Trp\n",
      "2025-06-03 19:29:20,828 - DEBUG - Found perfect candidate for gene-cvarcp_00013: tRNA-Pro\n",
      "2025-06-03 19:29:20,829 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00014\n",
      "2025-06-03 19:29:20,831 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00015\n",
      "2025-06-03 19:29:20,832 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00016\n",
      "2025-06-03 19:29:20,834 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00017\n",
      "2025-06-03 19:29:20,839 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00018\n",
      "2025-06-03 19:29:20,841 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00019\n",
      "2025-06-03 19:29:20,843 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00020\n",
      "2025-06-03 19:29:20,846 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00021\n",
      "2025-06-03 19:29:20,848 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00022\n",
      "2025-06-03 19:29:20,849 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00023\n",
      "2025-06-03 19:29:20,851 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00024\n",
      "2025-06-03 19:29:20,855 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00025\n",
      "2025-06-03 19:29:20,857 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00026\n",
      "2025-06-03 19:29:20,858 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00027\n",
      "2025-06-03 19:29:20,860 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00028\n",
      "2025-06-03 19:29:20,862 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00029\n",
      "2025-06-03 19:29:20,863 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00030\n",
      "2025-06-03 19:29:20,864 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00031\n",
      "2025-06-03 19:29:20,865 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00032\n",
      "2025-06-03 19:29:20,868 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00033\n",
      "2025-06-03 19:29:20,871 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00034\n",
      "2025-06-03 19:29:20,872 - DEBUG - Found perfect candidate for gene-cvarcp_00035: tRNA-Arg\n",
      "2025-06-03 19:29:20,874 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00036\n",
      "2025-06-03 19:29:20,875 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00037\n",
      "2025-06-03 19:29:20,877 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00038\n",
      "2025-06-03 19:29:20,879 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00039\n",
      "2025-06-03 19:29:20,880 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00040\n",
      "2025-06-03 19:29:20,881 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00041\n",
      "2025-06-03 19:29:20,883 - DEBUG - Found perfect candidate for gene-cvarcp_00042: tRNA-Arg\n",
      "2025-06-03 19:29:20,886 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00043\n",
      "2025-06-03 19:29:20,887 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00044\n",
      "2025-06-03 19:29:20,888 - DEBUG - Found perfect candidate for gene-cvarcp_00045: tRNA-Thr\n",
      "2025-06-03 19:29:20,890 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00046\n",
      "2025-06-03 19:29:20,891 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00047\n",
      "2025-06-03 19:29:20,893 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00048\n",
      "2025-06-03 19:29:20,894 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00049\n",
      "2025-06-03 19:29:20,896 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00050\n",
      "2025-06-03 19:29:20,897 - DEBUG - Found perfect candidate for gene-cvarcp_00051: tRNA-Met\n",
      "2025-06-03 19:29:20,898 - DEBUG - Found perfect candidate for gene-cvarcp_00053: tRNA-Asp\n",
      "2025-06-03 19:29:20,901 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00054\n",
      "2025-06-03 19:29:20,902 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00055\n",
      "2025-06-03 19:29:20,903 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00056\n",
      "2025-06-03 19:29:20,904 - DEBUG - Found perfect candidate for gene-cvarcp_00057: tRNA-Ser\n",
      "2025-06-03 19:29:20,911 - DEBUG - Found perfect candidate for gene-cvarcp_00058: tRNA-Leu\n",
      "2025-06-03 19:29:20,915 - DEBUG - Found perfect candidate for gene-cvarcp_00059: 16S ribosomal RNA\n",
      "2025-06-03 19:29:20,918 - DEBUG - Found perfect candidate for gene-cvarcp_00060: tRNA-Ile\n",
      "2025-06-03 19:29:20,921 - DEBUG - Found perfect candidate for gene-cvarcp_00061: tRNA-Ala\n",
      "2025-06-03 19:29:20,923 - DEBUG - Found perfect candidate for gene-cvarcp_00062: 23S ribosomal RNA\n",
      "2025-06-03 19:29:20,925 - DEBUG - Found perfect candidate for gene-cvarcp_00063: 5S ribosomal RNA\n",
      "2025-06-03 19:29:20,926 - DEBUG - Found perfect candidate for gene-cvarcp_00064: tRNA-Leu\n",
      "2025-06-03 19:29:20,928 - DEBUG - Found perfect candidate for gene-cvarcp_00065: tRNA-Val\n",
      "2025-06-03 19:29:20,931 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00066\n",
      "2025-06-03 19:29:20,934 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00067\n",
      "2025-06-03 19:29:20,936 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00068\n",
      "2025-06-03 19:29:20,938 - DEBUG - Found perfect candidate for gene-cvarcp_00069: tRNA-Thr\n",
      "2025-06-03 19:29:20,940 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00070\n",
      "2025-06-03 19:29:20,941 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00071\n",
      "2025-06-03 19:29:20,942 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00072\n",
      "2025-06-03 19:29:20,947 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00073\n",
      "2025-06-03 19:29:20,951 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00074\n",
      "2025-06-03 19:29:20,953 - DEBUG - Found perfect candidate for gene-cvarcp_00075: tRNA-Leu\n",
      "2025-06-03 19:29:20,955 - DEBUG - Found perfect candidate for gene-cvarcp_00076: tRNA-His\n",
      "2025-06-03 19:29:20,957 - DEBUG - Found perfect candidate for gene-cvarcp_00077: tRNA-Gly\n",
      "2025-06-03 19:29:20,958 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00078\n",
      "2025-06-03 19:29:20,960 - DEBUG - Found perfect candidate for gene-cvarcp_00079: tRNA-Lys\n",
      "2025-06-03 19:29:20,962 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00080\n",
      "2025-06-03 19:29:20,964 - DEBUG - Found perfect candidate for gene-cvarcp_00081: tRNA-Phe\n",
      "2025-06-03 19:29:20,965 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00082\n",
      "2025-06-03 19:29:20,967 - DEBUG - Found perfect candidate for gene-cvarcp_00083: tRNA-Gln\n",
      "2025-06-03 19:29:20,969 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00084\n",
      "2025-06-03 19:29:20,972 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00085\n",
      "2025-06-03 19:29:20,973 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00086\n",
      "2025-06-03 19:29:20,975 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00087\n",
      "2025-06-03 19:29:20,976 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00088\n",
      "2025-06-03 19:29:20,980 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00089\n",
      "2025-06-03 19:29:20,982 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00090\n",
      "2025-06-03 19:29:20,984 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00091\n",
      "2025-06-03 19:29:20,985 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00092\n",
      "2025-06-03 19:29:20,987 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00093\n",
      "2025-06-03 19:29:20,988 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00094\n",
      "2025-06-03 19:29:20,990 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00095\n",
      "2025-06-03 19:29:20,991 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00096\n",
      "2025-06-03 19:29:20,994 - DEBUG - Found perfect candidate for gene-cvarcp_00097: tRNA-Arg\n",
      "2025-06-03 19:29:20,997 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00098\n",
      "2025-06-03 19:29:21,001 - DEBUG - Found perfect candidate for gene-cvarcp_00099: tRNA-Asn\n",
      "2025-06-03 19:29:21,003 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00100\n",
      "2025-06-03 19:29:21,005 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00101\n",
      "2025-06-03 19:29:21,007 - DEBUG - Found perfect candidate for gene-cvarcp_00102: tRNA-Met\n",
      "2025-06-03 19:29:21,010 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00103\n",
      "2025-06-03 19:29:21,014 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00104\n",
      "2025-06-03 19:29:21,015 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00105\n",
      "2025-06-03 19:29:21,016 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00106\n",
      "2025-06-03 19:29:21,017 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00107\n",
      "2025-06-03 19:29:21,021 - DEBUG - Found perfect candidate for gene-cvarcp_00108: tRNA-Cys\n",
      "2025-06-03 19:29:21,022 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00109\n",
      "2025-06-03 19:29:21,024 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00110\n",
      "2025-06-03 19:29:21,026 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00111\n",
      "2025-06-03 19:29:21,027 - WARNING - no feature* found in KP271969.1 that have parent equals to gene-cvarcp_00112\n",
      "2025-06-03 19:29:21,029 - DEBUG - Found perfect candidate for gene-cvarcp_00113: tRNA-Leu\n",
      "2025-06-03 19:29:21,032 - DEBUG - Found perfect candidate for gene-cvarcp_00114: tRNA-Ser\n",
      "2025-06-03 19:29:21,034 - WARNING - no feature* found in LC523992.1 that have parent equals to gene-intronic orf281-2\n",
      "2025-06-03 19:29:21,036 - WARNING - no feature* found in LC523992.1 that have parent equals to gene-intronic orf281\n",
      "2025-06-03 19:29:21,037 - WARNING - no feature* found in LC523992.1 that have parent equals to gene-intronic orf307\n",
      "2025-06-03 19:29:21,039 - WARNING - no feature* found in KX756229.1 that have parent equals to gene-ltrA\n",
      "2025-06-03 19:29:21,041 - WARNING - no feature* found in KM462870.1 that have parent equals to gene-ssrA\n",
      "2025-06-03 19:29:21,041 - INFO - features* are those selected in the global_query_string variable, set at the beginning of the script\n"
     ]
    }
   ],
   "source": [
    "log.debug(\n",
    "    \"missing_dbxref: matching probable 'child feature + parent gene' pair (1st pass, matching rows of known features of features_info.tsv)\"\n",
    ")\n",
    "for row in features_unk_df.itertuples():\n",
    "    # check if there's a row in features_info_df with the same seqid and gene_id == parent\n",
    "    candidates = features_know_df[\n",
    "        (features_know_df[\"seqid\"] == row.seqid)\n",
    "        & (features_know_df[\"parent\"] == row.gene_id)\n",
    "    ]\n",
    "\n",
    "    if len(candidates) == 0:\n",
    "        log.warning(\n",
    "            f\"no feature* found in {row.seqid} that have parent equals to {row.gene_id}\"\n",
    "        )\n",
    "        continue\n",
    "\n",
    "    if len(candidates) > 1:\n",
    "        log.warning(\n",
    "            f\"more than one feature* found in {row.seqid} that have parent equals to {row.gene_id}\"\n",
    "        )\n",
    "\n",
    "        log.debug(\"Please check the candidates:\")\n",
    "        print_df_rows_log(candidates, log)\n",
    "        log.debug(\"Chosing the first one for feature_child\")\n",
    "        features_unk_df.at[row.Index, \"feature_child\"] = candidates.iloc[0][\n",
    "            \"feature_name\"\n",
    "        ]\n",
    "        continue\n",
    "\n",
    "    if len(candidates) == 1:\n",
    "        log.debug(\n",
    "            f'Found perfect candidate for {row.gene_id}: {candidates.iloc[0][\"feature_name\"]}'\n",
    "        )\n",
    "        features_unk_df.at[row.Index, \"feature_child\"] = candidates.iloc[0][\n",
    "            \"feature_name\"\n",
    "        ]\n",
    "\n",
    "log.info(\n",
    "    \"features* are those selected in the global_query_string variable, set at the beginning of the script\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "874acac7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 363 entries, 0 to 362\n",
      "Data columns (total 9 columns):\n",
      " #   Column         Non-Null Count  Dtype \n",
      "---  ------         --------------  ----- \n",
      " 0   seqid          363 non-null    object\n",
      " 1   gene_id        363 non-null    object\n",
      " 2   name           363 non-null    object\n",
      " 3   product        0 non-null      object\n",
      " 4   gene           7 non-null      object\n",
      " 5   parent         0 non-null      object\n",
      " 6   note           0 non-null      object\n",
      " 7   feature_name   363 non-null    object\n",
      " 8   feature_child  65 non-null     object\n",
      "dtypes: object(9)\n",
      "memory usage: 25.6+ KB\n"
     ]
    }
   ],
   "source": [
    "features_unk_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2a878ca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['12S ribosomal RNA', 'tRNA-Cys', 'tRNA-Arg', 'tRNA-Asn',\n",
       "       'tRNA-Tyr', 'tRNA-Lys', 'tRNA-Gly', 'tRNA-Asp', 'tRNA-Ser',\n",
       "       'tRNA-Trp', 'tRNA-Ile', 'tRNA-Pro', 'tRNA-Thr', 'tRNA-Glu',\n",
       "       'tRNA-Val', 'tRNA-Met', 'tRNA-Leu', 'tRNA-Ala', 'tRNA-Phe',\n",
       "       'tRNA-Gln', 'tRNA-His', 'small subunit ribosomal RNA', 'tRNA-Sec',\n",
       "       'large subunit ribosomal RNA', '12S RNA', '16S RNA', 'rrnS',\n",
       "       'rrnL', '23S ribosomal RNA', '16S ribosomal RNA'], dtype=object)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_child_uniques = features_unk_df[\"feature_child\"].dropna().unique()\n",
    "feature_child_uniques"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb3bb667",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: 12S ribosomal RNA not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Cys not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Arg not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Asn not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Tyr not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Lys not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Gly not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Asp not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Ser not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Trp not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Ile not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Pro not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Thr not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Glu not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Val not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Met not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Leu not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Ala not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Phe not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Gln not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-His not in gene_dict! how? did you run the step above without error?\n",
      "Error: small subunit ribosomal RNA not in gene_dict! how? did you run the step above without error?\n",
      "Error: tRNA-Sec not in gene_dict! how? did you run the step above without error?\n",
      "Error: large subunit ribosomal RNA not in gene_dict! how? did you run the step above without error?\n",
      "Error: 12S RNA not in gene_dict! how? did you run the step above without error?\n",
      "Error: 16S RNA not in gene_dict! how? did you run the step above without error?\n",
      "Error: rrnS not in gene_dict! how? did you run the step above without error?\n",
      "Error: rrnL not in gene_dict! how? did you run the step above without error?\n",
      "Error: 23S ribosomal RNA not in gene_dict! how? did you run the step above without error?\n",
      "Error: 16S ribosomal RNA not in gene_dict! how? did you run the step above without error?\n"
     ]
    }
   ],
   "source": [
    "# check if they exist in the gene_dict, they should if the script is correct\n",
    "log.debug(\"Checking if feature_childs are in gene_dict\")\n",
    "for feature_child in feature_child_uniques:\n",
    "    if feature_child not in gene_dict:\n",
    "        log.error(\n",
    "            f\"{feature_child} not in gene_dict! how? did you run the step above without error?\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f9a6d4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "comment = \"automated insertion from missing_dbxref_feature_child\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55aeab6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in features_unk_df[features_unk_df[\"feature_child\"].notna()].itertuples():\n",
    "    # sanity check\n",
    "    if r.feature_child not in gene_dict:\n",
    "        raise ValueError(\n",
    "            f\"Error: {r.feature_child} not in gene_dict! how? did you run the step above without error?\"\n",
    "        )\n",
    "\n",
    "    gene_dict[gene_id] = gdt.gene_dict.GeneGeneric(\n",
    "        label=gene_dict[r.feature_child].label, an_sources=r.seqid, c=comment\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5abed8ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO change header with number from increment_gdt_file\n",
    "most_recent_gdt_file, _ = increment_gdt_file(most_recent_gdt_file)\n",
    "gdt.gene_dict.write_gdt_file(gene_dict, most_recent_gdt_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9089390",
   "metadata": {},
   "source": [
    "#### unk_problems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f2083e55",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_unk = set()\n",
    "with open(MISC_DIR / \"features_unk.txt\", \"r\") as f1:\n",
    "    for line in f1:\n",
    "        line = line.strip()\n",
    "        if not line:\n",
    "            continue\n",
    "\n",
    "        if \"#gd\" in line:\n",
    "            line = line.split(\"#gd\")[0].strip()\n",
    "\n",
    "        names_unk.add(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8d1d2199",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_info_df = pd.read_csv(MISC_DIR / \"features_info.tsv\", sep=\"\\t\")\n",
    "features_unk_df = (\n",
    "    features_info_df[features_info_df[\"feature_name\"].isin(names_unk)]\n",
    "    .copy()\n",
    "    .reset_index(drop=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9cf8a50e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>name</th>\n",
       "      <th>product</th>\n",
       "      <th>gene</th>\n",
       "      <th>parent</th>\n",
       "      <th>note</th>\n",
       "      <th>feature_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0010</td>\n",
       "      <td>AFUA_m0010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0020</td>\n",
       "      <td>AFUA_m0020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0030</td>\n",
       "      <td>AFUA_m0030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0040</td>\n",
       "      <td>AFUA_m0040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0050</td>\n",
       "      <td>AFUA_m0050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0060</td>\n",
       "      <td>AFUA_m0060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0070</td>\n",
       "      <td>AFUA_m0070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0080</td>\n",
       "      <td>AFUA_m0080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0090</td>\n",
       "      <td>AFUA_m0090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0100</td>\n",
       "      <td>AFUA_m0100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0110</td>\n",
       "      <td>AFUA_m0110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0120</td>\n",
       "      <td>AFUA_m0120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0130</td>\n",
       "      <td>AFUA_m0130</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0140</td>\n",
       "      <td>AFUA_m0140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0150</td>\n",
       "      <td>AFUA_m0150</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0160</td>\n",
       "      <td>AFUA_m0160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0170</td>\n",
       "      <td>AFUA_m0170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0180</td>\n",
       "      <td>AFUA_m0180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0190</td>\n",
       "      <td>AFUA_m0190</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0200</td>\n",
       "      <td>AFUA_m0200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0210</td>\n",
       "      <td>AFUA_m0210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0220</td>\n",
       "      <td>AFUA_m0220</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0230</td>\n",
       "      <td>AFUA_m0230</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0240</td>\n",
       "      <td>AFUA_m0240</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0250</td>\n",
       "      <td>AFUA_m0250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0260</td>\n",
       "      <td>AFUA_m0260</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0270</td>\n",
       "      <td>AFUA_m0270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0280</td>\n",
       "      <td>AFUA_m0280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0290</td>\n",
       "      <td>AFUA_m0290</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0300</td>\n",
       "      <td>AFUA_m0300</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0310</td>\n",
       "      <td>AFUA_m0310</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0320</td>\n",
       "      <td>AFUA_m0320</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0330</td>\n",
       "      <td>AFUA_m0330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0340</td>\n",
       "      <td>AFUA_m0340</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0350</td>\n",
       "      <td>AFUA_m0350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0360</td>\n",
       "      <td>AFUA_m0360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0370</td>\n",
       "      <td>AFUA_m0370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0380</td>\n",
       "      <td>AFUA_m0380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0390</td>\n",
       "      <td>AFUA_m0390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0400</td>\n",
       "      <td>AFUA_m0400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0410</td>\n",
       "      <td>AFUA_m0410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0420</td>\n",
       "      <td>AFUA_m0420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0430</td>\n",
       "      <td>AFUA_m0430</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0440</td>\n",
       "      <td>AFUA_m0440</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0450</td>\n",
       "      <td>AFUA_m0450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0460</td>\n",
       "      <td>AFUA_m0460</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0470</td>\n",
       "      <td>AFUA_m0470</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0480</td>\n",
       "      <td>AFUA_m0480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0490</td>\n",
       "      <td>AFUA_m0490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0500</td>\n",
       "      <td>AFUA_m0500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0510</td>\n",
       "      <td>AFUA_m0510</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0520</td>\n",
       "      <td>AFUA_m0520</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>JQ346808.1</td>\n",
       "      <td>gene-AFUA_m0530</td>\n",
       "      <td>AFUA_m0530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AFUA_m0530</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqid          gene_id        name product gene parent note  \\\n",
       "0   JQ346808.1  gene-AFUA_m0010  AFUA_m0010     NaN  NaN    NaN  NaN   \n",
       "1   JQ346808.1  gene-AFUA_m0020  AFUA_m0020     NaN  NaN    NaN  NaN   \n",
       "2   JQ346808.1  gene-AFUA_m0030  AFUA_m0030     NaN  NaN    NaN  NaN   \n",
       "3   JQ346808.1  gene-AFUA_m0040  AFUA_m0040     NaN  NaN    NaN  NaN   \n",
       "4   JQ346808.1  gene-AFUA_m0050  AFUA_m0050     NaN  NaN    NaN  NaN   \n",
       "5   JQ346808.1  gene-AFUA_m0060  AFUA_m0060     NaN  NaN    NaN  NaN   \n",
       "6   JQ346808.1  gene-AFUA_m0070  AFUA_m0070     NaN  NaN    NaN  NaN   \n",
       "7   JQ346808.1  gene-AFUA_m0080  AFUA_m0080     NaN  NaN    NaN  NaN   \n",
       "8   JQ346808.1  gene-AFUA_m0090  AFUA_m0090     NaN  NaN    NaN  NaN   \n",
       "9   JQ346808.1  gene-AFUA_m0100  AFUA_m0100     NaN  NaN    NaN  NaN   \n",
       "10  JQ346808.1  gene-AFUA_m0110  AFUA_m0110     NaN  NaN    NaN  NaN   \n",
       "11  JQ346808.1  gene-AFUA_m0120  AFUA_m0120     NaN  NaN    NaN  NaN   \n",
       "12  JQ346808.1  gene-AFUA_m0130  AFUA_m0130     NaN  NaN    NaN  NaN   \n",
       "13  JQ346808.1  gene-AFUA_m0140  AFUA_m0140     NaN  NaN    NaN  NaN   \n",
       "14  JQ346808.1  gene-AFUA_m0150  AFUA_m0150     NaN  NaN    NaN  NaN   \n",
       "15  JQ346808.1  gene-AFUA_m0160  AFUA_m0160     NaN  NaN    NaN  NaN   \n",
       "16  JQ346808.1  gene-AFUA_m0170  AFUA_m0170     NaN  NaN    NaN  NaN   \n",
       "17  JQ346808.1  gene-AFUA_m0180  AFUA_m0180     NaN  NaN    NaN  NaN   \n",
       "18  JQ346808.1  gene-AFUA_m0190  AFUA_m0190     NaN  NaN    NaN  NaN   \n",
       "19  JQ346808.1  gene-AFUA_m0200  AFUA_m0200     NaN  NaN    NaN  NaN   \n",
       "20  JQ346808.1  gene-AFUA_m0210  AFUA_m0210     NaN  NaN    NaN  NaN   \n",
       "21  JQ346808.1  gene-AFUA_m0220  AFUA_m0220     NaN  NaN    NaN  NaN   \n",
       "22  JQ346808.1  gene-AFUA_m0230  AFUA_m0230     NaN  NaN    NaN  NaN   \n",
       "23  JQ346808.1  gene-AFUA_m0240  AFUA_m0240     NaN  NaN    NaN  NaN   \n",
       "24  JQ346808.1  gene-AFUA_m0250  AFUA_m0250     NaN  NaN    NaN  NaN   \n",
       "25  JQ346808.1  gene-AFUA_m0260  AFUA_m0260     NaN  NaN    NaN  NaN   \n",
       "26  JQ346808.1  gene-AFUA_m0270  AFUA_m0270     NaN  NaN    NaN  NaN   \n",
       "27  JQ346808.1  gene-AFUA_m0280  AFUA_m0280     NaN  NaN    NaN  NaN   \n",
       "28  JQ346808.1  gene-AFUA_m0290  AFUA_m0290     NaN  NaN    NaN  NaN   \n",
       "29  JQ346808.1  gene-AFUA_m0300  AFUA_m0300     NaN  NaN    NaN  NaN   \n",
       "30  JQ346808.1  gene-AFUA_m0310  AFUA_m0310     NaN  NaN    NaN  NaN   \n",
       "31  JQ346808.1  gene-AFUA_m0320  AFUA_m0320     NaN  NaN    NaN  NaN   \n",
       "32  JQ346808.1  gene-AFUA_m0330  AFUA_m0330     NaN  NaN    NaN  NaN   \n",
       "33  JQ346808.1  gene-AFUA_m0340  AFUA_m0340     NaN  NaN    NaN  NaN   \n",
       "34  JQ346808.1  gene-AFUA_m0350  AFUA_m0350     NaN  NaN    NaN  NaN   \n",
       "35  JQ346808.1  gene-AFUA_m0360  AFUA_m0360     NaN  NaN    NaN  NaN   \n",
       "36  JQ346808.1  gene-AFUA_m0370  AFUA_m0370     NaN  NaN    NaN  NaN   \n",
       "37  JQ346808.1  gene-AFUA_m0380  AFUA_m0380     NaN  NaN    NaN  NaN   \n",
       "38  JQ346808.1  gene-AFUA_m0390  AFUA_m0390     NaN  NaN    NaN  NaN   \n",
       "39  JQ346808.1  gene-AFUA_m0400  AFUA_m0400     NaN  NaN    NaN  NaN   \n",
       "40  JQ346808.1  gene-AFUA_m0410  AFUA_m0410     NaN  NaN    NaN  NaN   \n",
       "41  JQ346808.1  gene-AFUA_m0420  AFUA_m0420     NaN  NaN    NaN  NaN   \n",
       "42  JQ346808.1  gene-AFUA_m0430  AFUA_m0430     NaN  NaN    NaN  NaN   \n",
       "43  JQ346808.1  gene-AFUA_m0440  AFUA_m0440     NaN  NaN    NaN  NaN   \n",
       "44  JQ346808.1  gene-AFUA_m0450  AFUA_m0450     NaN  NaN    NaN  NaN   \n",
       "45  JQ346808.1  gene-AFUA_m0460  AFUA_m0460     NaN  NaN    NaN  NaN   \n",
       "46  JQ346808.1  gene-AFUA_m0470  AFUA_m0470     NaN  NaN    NaN  NaN   \n",
       "47  JQ346808.1  gene-AFUA_m0480  AFUA_m0480     NaN  NaN    NaN  NaN   \n",
       "48  JQ346808.1  gene-AFUA_m0490  AFUA_m0490     NaN  NaN    NaN  NaN   \n",
       "49  JQ346808.1  gene-AFUA_m0500  AFUA_m0500     NaN  NaN    NaN  NaN   \n",
       "50  JQ346808.1  gene-AFUA_m0510  AFUA_m0510     NaN  NaN    NaN  NaN   \n",
       "51  JQ346808.1  gene-AFUA_m0520  AFUA_m0520     NaN  NaN    NaN  NaN   \n",
       "52  JQ346808.1  gene-AFUA_m0530  AFUA_m0530     NaN  NaN    NaN  NaN   \n",
       "\n",
       "   feature_name  \n",
       "0    AFUA_m0010  \n",
       "1    AFUA_m0020  \n",
       "2    AFUA_m0030  \n",
       "3    AFUA_m0040  \n",
       "4    AFUA_m0050  \n",
       "5    AFUA_m0060  \n",
       "6    AFUA_m0070  \n",
       "7    AFUA_m0080  \n",
       "8    AFUA_m0090  \n",
       "9    AFUA_m0100  \n",
       "10   AFUA_m0110  \n",
       "11   AFUA_m0120  \n",
       "12   AFUA_m0130  \n",
       "13   AFUA_m0140  \n",
       "14   AFUA_m0150  \n",
       "15   AFUA_m0160  \n",
       "16   AFUA_m0170  \n",
       "17   AFUA_m0180  \n",
       "18   AFUA_m0190  \n",
       "19   AFUA_m0200  \n",
       "20   AFUA_m0210  \n",
       "21   AFUA_m0220  \n",
       "22   AFUA_m0230  \n",
       "23   AFUA_m0240  \n",
       "24   AFUA_m0250  \n",
       "25   AFUA_m0260  \n",
       "26   AFUA_m0270  \n",
       "27   AFUA_m0280  \n",
       "28   AFUA_m0290  \n",
       "29   AFUA_m0300  \n",
       "30   AFUA_m0310  \n",
       "31   AFUA_m0320  \n",
       "32   AFUA_m0330  \n",
       "33   AFUA_m0340  \n",
       "34   AFUA_m0350  \n",
       "35   AFUA_m0360  \n",
       "36   AFUA_m0370  \n",
       "37   AFUA_m0380  \n",
       "38   AFUA_m0390  \n",
       "39   AFUA_m0400  \n",
       "40   AFUA_m0410  \n",
       "41   AFUA_m0420  \n",
       "42   AFUA_m0430  \n",
       "43   AFUA_m0440  \n",
       "44   AFUA_m0450  \n",
       "45   AFUA_m0460  \n",
       "46   AFUA_m0470  \n",
       "47   AFUA_m0480  \n",
       "48   AFUA_m0490  \n",
       "49   AFUA_m0500  \n",
       "50   AFUA_m0510  \n",
       "51   AFUA_m0520  \n",
       "52   AFUA_m0530  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_unk_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e88a7ebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "unk_dict = features_unk_df.groupby(\"seqid\")[\"gene_id\"].agg(list).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1325b91c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'JQ346808.1': ['gene-AFUA_m0010',\n",
       "  'gene-AFUA_m0020',\n",
       "  'gene-AFUA_m0030',\n",
       "  'gene-AFUA_m0040',\n",
       "  'gene-AFUA_m0050',\n",
       "  'gene-AFUA_m0060',\n",
       "  'gene-AFUA_m0070',\n",
       "  'gene-AFUA_m0080',\n",
       "  'gene-AFUA_m0090',\n",
       "  'gene-AFUA_m0100',\n",
       "  'gene-AFUA_m0110',\n",
       "  'gene-AFUA_m0120',\n",
       "  'gene-AFUA_m0130',\n",
       "  'gene-AFUA_m0140',\n",
       "  'gene-AFUA_m0150',\n",
       "  'gene-AFUA_m0160',\n",
       "  'gene-AFUA_m0170',\n",
       "  'gene-AFUA_m0180',\n",
       "  'gene-AFUA_m0190',\n",
       "  'gene-AFUA_m0200',\n",
       "  'gene-AFUA_m0210',\n",
       "  'gene-AFUA_m0220',\n",
       "  'gene-AFUA_m0230',\n",
       "  'gene-AFUA_m0240',\n",
       "  'gene-AFUA_m0250',\n",
       "  'gene-AFUA_m0260',\n",
       "  'gene-AFUA_m0270',\n",
       "  'gene-AFUA_m0280',\n",
       "  'gene-AFUA_m0290',\n",
       "  'gene-AFUA_m0300',\n",
       "  'gene-AFUA_m0310',\n",
       "  'gene-AFUA_m0320',\n",
       "  'gene-AFUA_m0330',\n",
       "  'gene-AFUA_m0340',\n",
       "  'gene-AFUA_m0350',\n",
       "  'gene-AFUA_m0360',\n",
       "  'gene-AFUA_m0370',\n",
       "  'gene-AFUA_m0380',\n",
       "  'gene-AFUA_m0390',\n",
       "  'gene-AFUA_m0400',\n",
       "  'gene-AFUA_m0410',\n",
       "  'gene-AFUA_m0420',\n",
       "  'gene-AFUA_m0430',\n",
       "  'gene-AFUA_m0440',\n",
       "  'gene-AFUA_m0450',\n",
       "  'gene-AFUA_m0460',\n",
       "  'gene-AFUA_m0470',\n",
       "  'gene-AFUA_m0480',\n",
       "  'gene-AFUA_m0490',\n",
       "  'gene-AFUA_m0500',\n",
       "  'gene-AFUA_m0510',\n",
       "  'gene-AFUA_m0520',\n",
       "  'gene-AFUA_m0530']}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unk_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c7969df",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-06-06 18:29:22,018 - DEBUG - missing_dbxref: matching probable 'child feature + parent gene' pair (2nd pass, on the an original gff3, using all the features)\n",
      "2025-06-06 18:29:22,019 - DEBUG - AN: JQ346808.1| gene_ids: ['gene-AFUA_m0010', 'gene-AFUA_m0020', 'gene-AFUA_m0030', 'gene-AFUA_m0040', 'gene-AFUA_m0050', 'gene-AFUA_m0060', 'gene-AFUA_m0070', 'gene-AFUA_m0080', 'gene-AFUA_m0090', 'gene-AFUA_m0100', 'gene-AFUA_m0110', 'gene-AFUA_m0120', 'gene-AFUA_m0130', 'gene-AFUA_m0140', 'gene-AFUA_m0150', 'gene-AFUA_m0160', 'gene-AFUA_m0170', 'gene-AFUA_m0180', 'gene-AFUA_m0190', 'gene-AFUA_m0200', 'gene-AFUA_m0210', 'gene-AFUA_m0220', 'gene-AFUA_m0230', 'gene-AFUA_m0240', 'gene-AFUA_m0250', 'gene-AFUA_m0260', 'gene-AFUA_m0270', 'gene-AFUA_m0280', 'gene-AFUA_m0290', 'gene-AFUA_m0300', 'gene-AFUA_m0310', 'gene-AFUA_m0320', 'gene-AFUA_m0330', 'gene-AFUA_m0340', 'gene-AFUA_m0350', 'gene-AFUA_m0360', 'gene-AFUA_m0370', 'gene-AFUA_m0380', 'gene-AFUA_m0390', 'gene-AFUA_m0400', 'gene-AFUA_m0410', 'gene-AFUA_m0420', 'gene-AFUA_m0430', 'gene-AFUA_m0440', 'gene-AFUA_m0450', 'gene-AFUA_m0460', 'gene-AFUA_m0470', 'gene-AFUA_m0480', 'gene-AFUA_m0490', 'gene-AFUA_m0500', 'gene-AFUA_m0510', 'gene-AFUA_m0520', 'gene-AFUA_m0530']\n",
      "2025-06-06 18:29:22,028 - DEBUG -  gene-AFUA_m0010 | possible feature_name: cytochrome b\n",
      "2025-06-06 18:29:22,031 - DEBUG - \tt: CDS | p: cytochrome b | a: ID=cds-AFE02850.1;Parent=gene-AFUA_m0010;Dbxref=NCBI_GP:AFE02850.1;Name=AFE02850.1;gbkey=CDS;locus_tag=AFUA_m0010;product=cytochrome b;protein_id=AFE02850.1;transl_table=4\n",
      "2025-06-06 18:29:22,032 - DEBUG - \tlabel: MT-CYB, gene_dict[feature_name]: GeneDescription(label='MT-CYB', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,032 - DEBUG - \tgene-AFUA_m0010 | inserting from missing_dbxref_unk_problems, source: cytochrome b | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,034 - DEBUG -  gene-AFUA_m0020 | possible feature_name: tRNA-Cys\n",
      "2025-06-06 18:29:22,037 - DEBUG - \tt: tRNA | p: tRNA-Cys | a: ID=rna-AFUA_m0020;Parent=gene-AFUA_m0020;Note=anticodon GCA;gbkey=tRNA;locus_tag=AFUA_m0020;product=tRNA-Cys\n",
      "2025-06-06 18:29:22,037 - DEBUG - \tlabel: MT-TC, gene_dict[feature_name]: GeneDescription(label='MT-TC', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,038 - DEBUG - \tgene-AFUA_m0020 | inserting from missing_dbxref_unk_problems, source: tRNA-Cys | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,038 - DEBUG -  gene-AFUA_m0030 | possible feature_name: tRNA-Asn\n",
      "2025-06-06 18:29:22,040 - DEBUG - \tt: tRNA | p: tRNA-Asn | a: ID=rna-AFUA_m0030;Parent=gene-AFUA_m0030;Note=anticodon ATT;gbkey=tRNA;locus_tag=AFUA_m0030;product=tRNA-Asn\n",
      "2025-06-06 18:29:22,040 - DEBUG - \tlabel: MT-TN, gene_dict[feature_name]: GeneDescription(label='MT-TN', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,041 - DEBUG - \tgene-AFUA_m0030 | inserting from missing_dbxref_unk_problems, source: tRNA-Asn | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,042 - DEBUG -  gene-AFUA_m0040 | possible feature_name: NADH dehydrogenase subunit 1\n",
      "2025-06-06 18:29:22,045 - DEBUG - \tt: CDS | p: NADH dehydrogenase subunit 1 | a: ID=cds-AFE02851.1;Parent=gene-AFUA_m0040;Dbxref=NCBI_GP:AFE02851.1;Name=AFE02851.1;gbkey=CDS;locus_tag=AFUA_m0040;product=NADH dehydrogenase subunit 1;protein_id=AFE02851.1;transl_table=4\n",
      "2025-06-06 18:29:22,046 - DEBUG - \tlabel: MT-ND1, gene_dict[feature_name]: GeneDescription(label='MT-ND1', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,047 - DEBUG - \tgene-AFUA_m0040 | inserting from missing_dbxref_unk_problems, source: NADH dehydrogenase subunit 1 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,048 - DEBUG -  gene-AFUA_m0050 | possible feature_name: tRNA-Ile\n",
      "2025-06-06 18:29:22,049 - DEBUG - \tt: tRNA | p: tRNA-Ile | a: ID=rna-AFUA_m0050;Parent=gene-AFUA_m0050;Note=anticodon TAT;gbkey=tRNA;locus_tag=AFUA_m0050;product=tRNA-Ile\n",
      "2025-06-06 18:29:22,049 - DEBUG - \tlabel: MT-TI, gene_dict[feature_name]: GeneDescription(label='MT-TI', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,049 - DEBUG - \tgene-AFUA_m0050 | inserting from missing_dbxref_unk_problems, source: tRNA-Ile | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,050 - DEBUG -  gene-AFUA_m0060 | possible feature_name: NADH dehydrogenase subunit 4\n",
      "2025-06-06 18:29:22,052 - DEBUG - \tt: CDS | p: NADH dehydrogenase subunit 4 | a: ID=cds-AFE02852.1;Parent=gene-AFUA_m0060;Dbxref=NCBI_GP:AFE02852.1;Name=AFE02852.1;gbkey=CDS;locus_tag=AFUA_m0060;product=NADH dehydrogenase subunit 4;protein_id=AFE02852.1;transl_table=4\n",
      "2025-06-06 18:29:22,052 - DEBUG - \tlabel: MT-ND4, gene_dict[feature_name]: GeneDescription(label='MT-ND4', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,053 - DEBUG - \tgene-AFUA_m0060 | inserting from missing_dbxref_unk_problems, source: NADH dehydrogenase subunit 4 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,056 - DEBUG -  gene-AFUA_m0070 | possible feature_name: tRNA-Arg\n",
      "2025-06-06 18:29:22,057 - DEBUG - \tt: tRNA | p: tRNA-Arg | a: ID=rna-AFUA_m0070;Parent=gene-AFUA_m0070;Note=anticodon TCT;gbkey=tRNA;locus_tag=AFUA_m0070;product=tRNA-Arg\n",
      "2025-06-06 18:29:22,058 - DEBUG - \tlabel: MT-TR, gene_dict[feature_name]: GeneDescription(label='MT-TR', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,058 - DEBUG - \tgene-AFUA_m0070 | inserting from missing_dbxref_unk_problems, source: tRNA-Arg | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,059 - DEBUG -  gene-AFUA_m0080 | possible feature_name: tRNA-Asn\n",
      "2025-06-06 18:29:22,060 - DEBUG - \tt: tRNA | p: tRNA-Asn | a: ID=rna-AFUA_m0080;Parent=gene-AFUA_m0080;Note=anticodon GTT;gbkey=tRNA;locus_tag=AFUA_m0080;product=tRNA-Asn\n",
      "2025-06-06 18:29:22,061 - DEBUG - \tlabel: MT-TN, gene_dict[feature_name]: GeneDescription(label='MT-TN', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,061 - DEBUG - \tgene-AFUA_m0080 | inserting from missing_dbxref_unk_problems, source: tRNA-Asn | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,062 - DEBUG -  gene-AFUA_m0090 | possible feature_name: ATP synthase subunit 8\n",
      "2025-06-06 18:29:22,063 - DEBUG - \tt: CDS | p: ATP synthase subunit 8 | a: ID=cds-AFE02853.1;Parent=gene-AFUA_m0090;Dbxref=NCBI_GP:AFE02853.1;Name=AFE02853.1;gbkey=CDS;locus_tag=AFUA_m0090;product=ATP synthase subunit 8;protein_id=AFE02853.1;transl_table=4\n",
      "2025-06-06 18:29:22,064 - DEBUG - \tlabel: MT-ATP8, gene_dict[feature_name]: GeneDescription(label='MT-ATP8', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,064 - DEBUG - \tgene-AFUA_m0090 | inserting from missing_dbxref_unk_problems, source: ATP synthase subunit 8 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,065 - DEBUG -  gene-AFUA_m0100 | possible feature_name: tRNA-Ile\n",
      "2025-06-06 18:29:22,066 - DEBUG - \tt: tRNA | p: tRNA-Ile | a: ID=rna-AFUA_m0100;Parent=gene-AFUA_m0100;Note=anticodon TAT;gbkey=tRNA;locus_tag=AFUA_m0100;product=tRNA-Ile\n",
      "2025-06-06 18:29:22,067 - DEBUG - \tlabel: MT-TI, gene_dict[feature_name]: GeneDescription(label='MT-TI', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,067 - DEBUG - \tgene-AFUA_m0100 | inserting from missing_dbxref_unk_problems, source: tRNA-Ile | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,069 - DEBUG -  gene-AFUA_m0110 | possible feature_name: ATP synthase subunit 6\n",
      "2025-06-06 18:29:22,071 - DEBUG - \tt: CDS | p: ATP synthase subunit 6 | a: ID=cds-AFE02854.1;Parent=gene-AFUA_m0110;Dbxref=NCBI_GP:AFE02854.1;Name=AFE02854.1;gbkey=CDS;locus_tag=AFUA_m0110;product=ATP synthase subunit 6;protein_id=AFE02854.1;transl_table=4\n",
      "2025-06-06 18:29:22,071 - DEBUG - \tlabel: MT-ATP6, gene_dict[feature_name]: GeneDescription(label='MT-ATP6', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,072 - DEBUG - \tgene-AFUA_m0110 | inserting from missing_dbxref_unk_problems, source: ATP synthase subunit 6 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,073 - DEBUG -  gene-AFUA_m0120 | possible feature_name: small subunit ribosomal RNA\n",
      "2025-06-06 18:29:22,074 - DEBUG - \tt: rRNA | p: small subunit ribosomal RNA | a: ID=rna-AFUA_m0120;Parent=gene-AFUA_m0120;gbkey=rRNA;locus_tag=AFUA_m0120;product=small subunit ribosomal RNA\n",
      "2025-06-06 18:29:22,075 - DEBUG - \tlabel: MT-RNR1, gene_dict[feature_name]: GeneDescription(label='MT-RNR1', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,077 - DEBUG - \tgene-AFUA_m0120 | inserting from missing_dbxref_unk_problems, source: small subunit ribosomal RNA | type: rRNA\n",
      "\n",
      "2025-06-06 18:29:22,078 - DEBUG -  gene-AFUA_m0130 | possible feature_name: tRNA-Tyr\n",
      "2025-06-06 18:29:22,079 - DEBUG - \tt: tRNA | p: tRNA-Tyr | a: ID=rna-AFUA_m0130;Parent=gene-AFUA_m0130;Note=anticodon GTA;gbkey=tRNA;locus_tag=AFUA_m0130;product=tRNA-Tyr\n",
      "2025-06-06 18:29:22,080 - DEBUG - \tlabel: MT-TY, gene_dict[feature_name]: GeneDescription(label='MT-TY', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,080 - DEBUG - \tgene-AFUA_m0130 | inserting from missing_dbxref_unk_problems, source: tRNA-Tyr | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,081 - DEBUG -  gene-AFUA_m0140 | possible feature_name: NADH dehydrogenase subunit 6\n",
      "2025-06-06 18:29:22,082 - DEBUG - \tt: CDS | p: NADH dehydrogenase subunit 6 | a: ID=cds-AFE02855.1;Parent=gene-AFUA_m0140;Dbxref=NCBI_GP:AFE02855.1;Name=AFE02855.1;gbkey=CDS;locus_tag=AFUA_m0140;product=NADH dehydrogenase subunit 6;protein_id=AFE02855.1;transl_table=4\n",
      "2025-06-06 18:29:22,083 - DEBUG - \tlabel: MT-ND6, gene_dict[feature_name]: GeneDescription(label='MT-ND6', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,083 - DEBUG - \tgene-AFUA_m0140 | inserting from missing_dbxref_unk_problems, source: NADH dehydrogenase subunit 6 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,084 - DEBUG -  gene-AFUA_m0150 | possible feature_name: cytochrome oxidase subunit 3\n",
      "2025-06-06 18:29:22,086 - DEBUG - \tt: CDS | p: cytochrome oxidase subunit 3 | a: ID=cds-AFE02856.1;Parent=gene-AFUA_m0150;Dbxref=NCBI_GP:AFE02856.1;Name=AFE02856.1;gbkey=CDS;locus_tag=AFUA_m0150;product=cytochrome oxidase subunit 3;protein_id=AFE02856.1;transl_table=4\n",
      "2025-06-06 18:29:22,087 - DEBUG - \tlabel: MT-CO3, gene_dict[feature_name]: GeneDescription(label='MT-CO3', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,088 - DEBUG - \tgene-AFUA_m0150 | inserting from missing_dbxref_unk_problems, source: cytochrome oxidase subunit 3 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,089 - DEBUG -  gene-AFUA_m0160 | possible feature_name: tRNA-Lys\n",
      "2025-06-06 18:29:22,090 - DEBUG - \tt: tRNA | p: tRNA-Lys | a: ID=rna-AFUA_m0160;Parent=gene-AFUA_m0160;Note=anticodon TTT;gbkey=tRNA;locus_tag=AFUA_m0160;product=tRNA-Lys\n",
      "2025-06-06 18:29:22,090 - DEBUG - \tlabel: MT-TK, gene_dict[feature_name]: GeneDescription(label='MT-TK', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,090 - DEBUG - \tgene-AFUA_m0160 | inserting from missing_dbxref_unk_problems, source: tRNA-Lys | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,091 - DEBUG -  gene-AFUA_m0170 | possible feature_name: tRNA-Gly\n",
      "2025-06-06 18:29:22,092 - DEBUG - \tt: tRNA | p: tRNA-Gly | a: ID=rna-AFUA_m0170;Parent=gene-AFUA_m0170;Note=anticodon ACC;gbkey=tRNA;locus_tag=AFUA_m0170;product=tRNA-Gly\n",
      "2025-06-06 18:29:22,093 - DEBUG - \tlabel: MT-TG, gene_dict[feature_name]: GeneDescription(label='MT-TG', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,093 - DEBUG - \tgene-AFUA_m0170 | inserting from missing_dbxref_unk_problems, source: tRNA-Gly | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,094 - DEBUG -  gene-AFUA_m0180 | possible feature_name: tRNA-Gly\n",
      "2025-06-06 18:29:22,172 - DEBUG - \tt: tRNA | p: tRNA-Gly | a: ID=rna-AFUA_m0180;Parent=gene-AFUA_m0180;Note=anticodon TCC;gbkey=tRNA;locus_tag=AFUA_m0180;product=tRNA-Gly\n",
      "2025-06-06 18:29:22,173 - DEBUG - \tlabel: MT-TG, gene_dict[feature_name]: GeneDescription(label='MT-TG', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,174 - DEBUG - \tgene-AFUA_m0180 | inserting from missing_dbxref_unk_problems, source: tRNA-Gly | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,175 - DEBUG -  gene-AFUA_m0190 | possible feature_name: tRNA-Asp\n",
      "2025-06-06 18:29:22,176 - DEBUG - \tt: tRNA | p: tRNA-Asp | a: ID=rna-AFUA_m0190;Parent=gene-AFUA_m0190;Note=anticodon GTC;gbkey=tRNA;locus_tag=AFUA_m0190;product=tRNA-Asp\n",
      "2025-06-06 18:29:22,176 - DEBUG - \tlabel: MT-TD, gene_dict[feature_name]: GeneDescription(label='MT-TD', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,177 - DEBUG - \tgene-AFUA_m0190 | inserting from missing_dbxref_unk_problems, source: tRNA-Asp | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,178 - DEBUG -  gene-AFUA_m0200 | possible feature_name: tRNA-Ser\n",
      "2025-06-06 18:29:22,179 - DEBUG - \tt: tRNA | p: tRNA-Ser | a: ID=rna-AFUA_m0200;Parent=gene-AFUA_m0200;Note=anticodon GCT;gbkey=tRNA;locus_tag=AFUA_m0200;product=tRNA-Ser\n",
      "2025-06-06 18:29:22,179 - DEBUG - \tlabel: MT-TS1|2, gene_dict[feature_name]: GeneDescription(label='MT-TS1|2', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,180 - DEBUG - \tgene-AFUA_m0200 | inserting from missing_dbxref_unk_problems, source: tRNA-Ser | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,181 - DEBUG -  gene-AFUA_m0210 | possible feature_name: tRNA-Sec\n",
      "2025-06-06 18:29:22,182 - DEBUG - \tt: tRNA | p: tRNA-Sec | a: ID=rna-AFUA_m0210;Parent=gene-AFUA_m0210;Note=anticodon TCA;gbkey=tRNA;locus_tag=AFUA_m0210;product=tRNA-Sec\n",
      "2025-06-06 18:29:22,184 - DEBUG - \tlabel: MT-TSEC, gene_dict[feature_name]: GeneDescription(label='MT-TSEC', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,184 - DEBUG - \tgene-AFUA_m0210 | inserting from missing_dbxref_unk_problems, source: tRNA-Sec | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,186 - DEBUG -  gene-AFUA_m0220 | possible feature_name: tRNA-Ile\n",
      "2025-06-06 18:29:22,187 - DEBUG - \tt: tRNA | p: tRNA-Ile | a: ID=rna-AFUA_m0220;Parent=gene-AFUA_m0220;Note=anticodon GAT;gbkey=tRNA;locus_tag=AFUA_m0220;product=tRNA-Ile\n",
      "2025-06-06 18:29:22,189 - DEBUG - \tlabel: MT-TI, gene_dict[feature_name]: GeneDescription(label='MT-TI', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,190 - DEBUG - \tgene-AFUA_m0220 | inserting from missing_dbxref_unk_problems, source: tRNA-Ile | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,192 - DEBUG -  gene-AFUA_m0230 | possible feature_name: tRNA-Ser\n",
      "2025-06-06 18:29:22,194 - DEBUG - \tt: tRNA | p: tRNA-Ser | a: ID=rna-AFUA_m0230;Parent=gene-AFUA_m0230;Note=anticodon TGA;gbkey=tRNA;locus_tag=AFUA_m0230;product=tRNA-Ser\n",
      "2025-06-06 18:29:22,194 - DEBUG - \tlabel: MT-TS1|2, gene_dict[feature_name]: GeneDescription(label='MT-TS1|2', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,196 - DEBUG - \tgene-AFUA_m0230 | inserting from missing_dbxref_unk_problems, source: tRNA-Ser | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,198 - DEBUG -  gene-AFUA_m0240 | possible feature_name: tRNA-Pro\n",
      "2025-06-06 18:29:22,200 - DEBUG - \tt: tRNA | p: tRNA-Pro | a: ID=rna-AFUA_m0240;Parent=gene-AFUA_m0240;Note=anticodon TGG;gbkey=tRNA;locus_tag=AFUA_m0240;product=tRNA-Pro\n",
      "2025-06-06 18:29:22,201 - DEBUG - \tlabel: MT-TP, gene_dict[feature_name]: GeneDescription(label='MT-TP', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,201 - DEBUG - \tgene-AFUA_m0240 | inserting from missing_dbxref_unk_problems, source: tRNA-Pro | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,202 - DEBUG -  gene-AFUA_m0250 | possible feature_name: large subunit ribosomal RNA\n",
      "2025-06-06 18:29:22,204 - DEBUG - \tt: rRNA | p: large subunit ribosomal RNA | a: ID=rna-AFUA_m0250;Parent=gene-AFUA_m0250;gbkey=rRNA;locus_tag=AFUA_m0250;product=large subunit ribosomal RNA\n",
      "2025-06-06 18:29:22,205 - DEBUG - \tlabel: MT-RNR2, gene_dict[feature_name]: GeneDescription(label='MT-RNR2', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,205 - DEBUG - \tgene-AFUA_m0250 | inserting from missing_dbxref_unk_problems, source: large subunit ribosomal RNA | type: rRNA\n",
      "\n",
      "2025-06-06 18:29:22,206 - DEBUG -  gene-AFUA_m0260 | possible feature_name: ribosomal protein S5\n",
      "2025-06-06 18:29:22,207 - DEBUG - \tt: CDS | p: ribosomal protein S5 | a: ID=cds-AFE02857.1;Parent=gene-AFUA_m0260;Dbxref=NCBI_GP:AFE02857.1;Name=AFE02857.1;gbkey=CDS;locus_tag=AFUA_m0260;product=ribosomal protein S5;protein_id=AFE02857.1;transl_table=4\n",
      "2025-06-06 18:29:22,208 - DEBUG - \tlabel: MT-RPS5, gene_dict[feature_name]: GeneDescription(label='MT-RPS5', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,208 - DEBUG - \tgene-AFUA_m0260 | inserting from missing_dbxref_unk_problems, source: ribosomal protein S5 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,209 - DEBUG -  gene-AFUA_m0270 | possible feature_name: tRNA-Thr\n",
      "2025-06-06 18:29:22,211 - DEBUG - \tt: tRNA | p: tRNA-Thr | a: ID=rna-AFUA_m0270;Parent=gene-AFUA_m0270;Note=anticodon TGT;gbkey=tRNA;locus_tag=AFUA_m0270;product=tRNA-Thr\n",
      "2025-06-06 18:29:22,212 - DEBUG - \tlabel: MT-TT, gene_dict[feature_name]: GeneDescription(label='MT-TT', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,213 - DEBUG - \tgene-AFUA_m0270 | inserting from missing_dbxref_unk_problems, source: tRNA-Thr | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,214 - DEBUG -  gene-AFUA_m0280 | possible feature_name: tRNA-Glu\n",
      "2025-06-06 18:29:22,215 - DEBUG - \tt: tRNA | p: tRNA-Glu | a: ID=rna-AFUA_m0280;Parent=gene-AFUA_m0280;Note=anticodon TTC;gbkey=tRNA;locus_tag=AFUA_m0280;product=tRNA-Glu\n",
      "2025-06-06 18:29:22,215 - DEBUG - \tlabel: MT-TE, gene_dict[feature_name]: GeneDescription(label='MT-TE', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,216 - DEBUG - \tgene-AFUA_m0280 | inserting from missing_dbxref_unk_problems, source: tRNA-Glu | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,218 - DEBUG -  gene-AFUA_m0290 | possible feature_name: tRNA-Val\n",
      "2025-06-06 18:29:22,219 - DEBUG - \tt: tRNA | p: tRNA-Val | a: ID=rna-AFUA_m0290;Parent=gene-AFUA_m0290;Note=anticodon TAC;gbkey=tRNA;locus_tag=AFUA_m0290;product=tRNA-Val\n",
      "2025-06-06 18:29:22,219 - DEBUG - \tlabel: MT-TV, gene_dict[feature_name]: GeneDescription(label='MT-TV', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,220 - DEBUG - \tgene-AFUA_m0290 | inserting from missing_dbxref_unk_problems, source: tRNA-Val | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,221 - DEBUG -  gene-AFUA_m0300 | possible feature_name: tRNA-Met\n",
      "2025-06-06 18:29:22,223 - DEBUG - \tt: tRNA | p: tRNA-Met | a: ID=rna-AFUA_m0300;Parent=gene-AFUA_m0300;Note=anticodon CAT;gbkey=tRNA;locus_tag=AFUA_m0300;product=tRNA-Met\n",
      "2025-06-06 18:29:22,223 - DEBUG - \tlabel: MT-TM, gene_dict[feature_name]: GeneDescription(label='MT-TM', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,224 - DEBUG - \tgene-AFUA_m0300 | inserting from missing_dbxref_unk_problems, source: tRNA-Met | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,226 - DEBUG -  gene-AFUA_m0310 | possible feature_name: tRNA-Met\n",
      "2025-06-06 18:29:22,227 - DEBUG - \tt: tRNA | p: tRNA-Met | a: ID=rna-AFUA_m0310;Parent=gene-AFUA_m0310;Note=anticodon CAT;gbkey=tRNA;locus_tag=AFUA_m0310;product=tRNA-Met\n",
      "2025-06-06 18:29:22,228 - DEBUG - \tlabel: MT-TM, gene_dict[feature_name]: GeneDescription(label='MT-TM', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,228 - DEBUG - \tgene-AFUA_m0310 | inserting from missing_dbxref_unk_problems, source: tRNA-Met | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,230 - DEBUG -  gene-AFUA_m0320 | possible feature_name: tRNA-Leu\n",
      "2025-06-06 18:29:22,232 - DEBUG - \tt: tRNA | p: tRNA-Leu | a: ID=rna-AFUA_m0320;Parent=gene-AFUA_m0320;Note=anticodon TAA;gbkey=tRNA;locus_tag=AFUA_m0320;product=tRNA-Leu\n",
      "2025-06-06 18:29:22,232 - DEBUG - \tlabel: MT-TL1|2, gene_dict[feature_name]: GeneDescription(label='MT-TL1|2', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,233 - DEBUG - \tgene-AFUA_m0320 | inserting from missing_dbxref_unk_problems, source: tRNA-Leu | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,234 - DEBUG -  gene-AFUA_m0330 | possible feature_name: tRNA-Ala\n",
      "2025-06-06 18:29:22,235 - DEBUG - \tt: tRNA | p: tRNA-Ala | a: ID=rna-AFUA_m0330;Parent=gene-AFUA_m0330;Note=anticodon TGC;gbkey=tRNA;locus_tag=AFUA_m0330;product=tRNA-Ala\n",
      "2025-06-06 18:29:22,237 - DEBUG - \tlabel: MT-TA, gene_dict[feature_name]: GeneDescription(label='MT-TA', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,238 - DEBUG - \tgene-AFUA_m0330 | inserting from missing_dbxref_unk_problems, source: tRNA-Ala | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,239 - DEBUG -  gene-AFUA_m0340 | possible feature_name: tRNA-Phe\n",
      "2025-06-06 18:29:22,240 - DEBUG - \tt: tRNA | p: tRNA-Phe | a: ID=rna-AFUA_m0340;Parent=gene-AFUA_m0340;Note=anticodon GAA;gbkey=tRNA;locus_tag=AFUA_m0340;product=tRNA-Phe\n",
      "2025-06-06 18:29:22,243 - DEBUG - \tlabel: MT-TF, gene_dict[feature_name]: GeneDescription(label='MT-TF', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,244 - DEBUG - \tgene-AFUA_m0340 | inserting from missing_dbxref_unk_problems, source: tRNA-Phe | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,247 - DEBUG -  gene-AFUA_m0350 | possible feature_name: tRNA-Leu\n",
      "2025-06-06 18:29:22,248 - DEBUG - \tt: tRNA | p: tRNA-Leu | a: ID=rna-AFUA_m0350;Parent=gene-AFUA_m0350;Note=anticodon TAG;gbkey=tRNA;locus_tag=AFUA_m0350;product=tRNA-Leu\n",
      "2025-06-06 18:29:22,248 - DEBUG - \tlabel: MT-TL1|2, gene_dict[feature_name]: GeneDescription(label='MT-TL1|2', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,249 - DEBUG - \tgene-AFUA_m0350 | inserting from missing_dbxref_unk_problems, source: tRNA-Leu | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,249 - DEBUG -  gene-AFUA_m0360 | possible feature_name: tRNA-Gln\n",
      "2025-06-06 18:29:22,251 - DEBUG - \tt: tRNA | p: tRNA-Gln | a: ID=rna-AFUA_m0360;Parent=gene-AFUA_m0360;Note=anticodon TTG;gbkey=tRNA;locus_tag=AFUA_m0360;product=tRNA-Gln\n",
      "2025-06-06 18:29:22,251 - DEBUG - \tlabel: MT-TQ, gene_dict[feature_name]: GeneDescription(label='MT-TQ', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,252 - DEBUG - \tgene-AFUA_m0360 | inserting from missing_dbxref_unk_problems, source: tRNA-Gln | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,253 - DEBUG -  gene-AFUA_m0370 | possible feature_name: tRNA-Met\n",
      "2025-06-06 18:29:22,255 - DEBUG - \tt: tRNA | p: tRNA-Met | a: ID=rna-AFUA_m0370;Parent=gene-AFUA_m0370;Note=anticodon CAT;gbkey=tRNA;locus_tag=AFUA_m0370;product=tRNA-Met\n",
      "2025-06-06 18:29:22,256 - DEBUG - \tlabel: MT-TM, gene_dict[feature_name]: GeneDescription(label='MT-TM', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,256 - DEBUG - \tgene-AFUA_m0370 | inserting from missing_dbxref_unk_problems, source: tRNA-Met | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,257 - DEBUG -  gene-AFUA_m0380 | possible feature_name: tRNA-Ile\n",
      "2025-06-06 18:29:22,258 - DEBUG - \tt: tRNA | p: tRNA-Ile | a: ID=rna-AFUA_m0380;Parent=gene-AFUA_m0380;Note=anticodon TAT;gbkey=tRNA;locus_tag=AFUA_m0380;product=tRNA-Ile\n",
      "2025-06-06 18:29:22,259 - DEBUG - \tlabel: MT-TI, gene_dict[feature_name]: GeneDescription(label='MT-TI', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,259 - DEBUG - \tgene-AFUA_m0380 | inserting from missing_dbxref_unk_problems, source: tRNA-Ile | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,260 - DEBUG -  gene-AFUA_m0390 | possible feature_name: hypothetical protein\n",
      "2025-06-06 18:29:22,262 - DEBUG - \tt: CDS | p: hypothetical protein | a: ID=cds-AFE02858.1;Parent=gene-AFUA_m0390;Dbxref=NCBI_GP:AFE02858.1;Name=AFE02858.1;gbkey=CDS;locus_tag=AFUA_m0390;product=hypothetical protein;protein_id=AFE02858.1;transl_table=4\n",
      "2025-06-06 18:29:22,262 - DEBUG - \tproduct not in gene_dict: hypothetical protein\n",
      "\n",
      "2025-06-06 18:29:22,263 - DEBUG - \tchecking in temp_unk\n",
      "2025-06-06 18:29:22,263 - DEBUG - \tgene-AFUA_m0390 | not found in temp_unk, creating new entry\n",
      "2025-06-06 18:29:22,264 - DEBUG -  gene-AFUA_m0400 | possible feature_name: hypothetical protein\n",
      "2025-06-06 18:29:22,266 - DEBUG - \tt: CDS | p: hypothetical protein | a: ID=cds-AFE02859.1;Parent=gene-AFUA_m0400;Dbxref=NCBI_GP:AFE02859.1;Name=AFE02859.1;gbkey=CDS;locus_tag=AFUA_m0400;product=hypothetical protein;protein_id=AFE02859.1;transl_table=4\n",
      "2025-06-06 18:29:22,266 - DEBUG - \tproduct not in gene_dict: hypothetical protein\n",
      "\n",
      "2025-06-06 18:29:22,267 - DEBUG - \tchecking in temp_unk\n",
      "2025-06-06 18:29:22,267 - DEBUG - \tgene-AFUA_m0400 | found in temp_unk, inserting gene_id in there\n",
      "2025-06-06 18:29:22,268 - DEBUG -  gene-AFUA_m0410 | possible feature_name: tRNA-His\n",
      "2025-06-06 18:29:22,269 - DEBUG - \tt: tRNA | p: tRNA-His | a: ID=rna-AFUA_m0410;Parent=gene-AFUA_m0410;Note=anticodon GTG;gbkey=tRNA;locus_tag=AFUA_m0410;product=tRNA-His\n",
      "2025-06-06 18:29:22,270 - DEBUG - \tlabel: MT-TH, gene_dict[feature_name]: GeneDescription(label='MT-TH', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,270 - DEBUG - \tgene-AFUA_m0410 | inserting from missing_dbxref_unk_problems, source: tRNA-His | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,271 - DEBUG -  gene-AFUA_m0420 | possible feature_name: cytochrome oxidase subunit 1\n",
      "2025-06-06 18:29:22,272 - DEBUG - \tt: CDS | p: cytochrome oxidase subunit 1 | a: ID=cds-AFE02860.1;Parent=gene-AFUA_m0420;Dbxref=NCBI_GP:AFE02860.1;Name=AFE02860.1;gbkey=CDS;locus_tag=AFUA_m0420;product=cytochrome oxidase subunit 1;protein_id=AFE02860.1;transl_table=4\n",
      "2025-06-06 18:29:22,272 - DEBUG - \tt: CDS | p: cytochrome oxidase subunit 1 | a: ID=cds-AFE02860.1;Parent=gene-AFUA_m0420;Dbxref=NCBI_GP:AFE02860.1;Name=AFE02860.1;gbkey=CDS;locus_tag=AFUA_m0420;product=cytochrome oxidase subunit 1;protein_id=AFE02860.1;transl_table=4\n",
      "2025-06-06 18:29:22,273 - DEBUG - \tlabel: MT-CO1, gene_dict[feature_name]: GeneDescription(label='MT-CO1', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,273 - DEBUG - \tgene-AFUA_m0420 | inserting from missing_dbxref_unk_problems, source: cytochrome oxidase subunit 1 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,274 - DEBUG -  gene-AFUA_m0430 | possible feature_name: intron-encoded LAGLIDADG endonuclease family protein\n",
      "2025-06-06 18:29:22,275 - DEBUG - \tt: CDS | p: intron-encoded LAGLIDADG endonuclease family protein | a: ID=cds-AFE02861.1;Parent=gene-AFUA_m0430;Dbxref=NCBI_GP:AFE02861.1;Name=AFE02861.1;gbkey=CDS;locus_tag=AFUA_m0430;product=intron-encoded LAGLIDADG endonuclease family protein;protein_id=AFE02861.1;transl_table=4\n",
      "2025-06-06 18:29:22,276 - DEBUG - \tproduct not in gene_dict: intron-encoded LAGLIDADG endonuclease family protein\n",
      "\n",
      "2025-06-06 18:29:22,277 - DEBUG - \tchecking in temp_unk\n",
      "2025-06-06 18:29:22,277 - DEBUG - \tgene-AFUA_m0430 | not found in temp_unk, creating new entry\n",
      "2025-06-06 18:29:22,278 - DEBUG -  gene-AFUA_m0440 | possible feature_name: hypothetical protein\n",
      "2025-06-06 18:29:22,279 - DEBUG - \tt: CDS | p: hypothetical protein | a: ID=cds-AFE02862.1;Parent=gene-AFUA_m0440;Dbxref=NCBI_GP:AFE02862.1;Name=AFE02862.1;gbkey=CDS;locus_tag=AFUA_m0440;product=hypothetical protein;protein_id=AFE02862.1;transl_table=4\n",
      "2025-06-06 18:29:22,279 - DEBUG - \tproduct not in gene_dict: hypothetical protein\n",
      "\n",
      "2025-06-06 18:29:22,280 - DEBUG - \tchecking in temp_unk\n",
      "2025-06-06 18:29:22,280 - DEBUG - \tgene-AFUA_m0440 | found in temp_unk, inserting gene_id in there\n",
      "2025-06-06 18:29:22,281 - DEBUG -  gene-AFUA_m0450 | possible feature_name: hypothetical protein\n",
      "2025-06-06 18:29:22,282 - DEBUG - \tt: CDS | p: hypothetical protein | a: ID=cds-AFE02863.1;Parent=gene-AFUA_m0450;Dbxref=NCBI_GP:AFE02863.1;Name=AFE02863.1;gbkey=CDS;locus_tag=AFUA_m0450;product=hypothetical protein;protein_id=AFE02863.1;transl_table=4\n",
      "2025-06-06 18:29:22,282 - DEBUG - \tproduct not in gene_dict: hypothetical protein\n",
      "\n",
      "2025-06-06 18:29:22,283 - DEBUG - \tchecking in temp_unk\n",
      "2025-06-06 18:29:22,283 - DEBUG - \tgene-AFUA_m0450 | found in temp_unk, inserting gene_id in there\n",
      "2025-06-06 18:29:22,286 - DEBUG -  gene-AFUA_m0460 | possible feature_name: ATP synthase subunit 9\n",
      "2025-06-06 18:29:22,288 - DEBUG - \tt: CDS | p: ATP synthase subunit 9 | a: ID=cds-AFE02864.1;Parent=gene-AFUA_m0460;Dbxref=NCBI_GP:AFE02864.1;Name=AFE02864.1;gbkey=CDS;locus_tag=AFUA_m0460;product=ATP synthase subunit 9;protein_id=AFE02864.1;transl_table=4\n",
      "2025-06-06 18:29:22,289 - DEBUG - \tlabel: MT-ATP9, gene_dict[feature_name]: GeneDescription(label='MT-ATP9', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,289 - DEBUG - \tgene-AFUA_m0460 | inserting from missing_dbxref_unk_problems, source: ATP synthase subunit 9 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,291 - DEBUG -  gene-AFUA_m0470 | possible feature_name: tRNA-Ala\n",
      "2025-06-06 18:29:22,293 - DEBUG - \tt: tRNA | p: tRNA-Ala | a: ID=rna-AFUA_m0470;Parent=gene-AFUA_m0470;Note=anticodon TGC;gbkey=tRNA;locus_tag=AFUA_m0470;product=tRNA-Ala\n",
      "2025-06-06 18:29:22,293 - DEBUG - \tlabel: MT-TA, gene_dict[feature_name]: GeneDescription(label='MT-TA', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,293 - DEBUG - \tgene-AFUA_m0470 | inserting from missing_dbxref_unk_problems, source: tRNA-Ala | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,294 - DEBUG -  gene-AFUA_m0480 | possible feature_name: NADH dehydrogenase subunit 3\n",
      "2025-06-06 18:29:22,295 - DEBUG - \tt: CDS | p: NADH dehydrogenase subunit 3 | a: ID=cds-AFE02865.1;Parent=gene-AFUA_m0480;Dbxref=NCBI_GP:AFE02865.1;Name=AFE02865.1;gbkey=CDS;locus_tag=AFUA_m0480;product=NADH dehydrogenase subunit 3;protein_id=AFE02865.1;transl_table=4\n",
      "2025-06-06 18:29:22,296 - DEBUG - \tlabel: MT-ND3, gene_dict[feature_name]: GeneDescription(label='MT-ND3', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,296 - DEBUG - \tgene-AFUA_m0480 | inserting from missing_dbxref_unk_problems, source: NADH dehydrogenase subunit 3 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,297 - DEBUG -  gene-AFUA_m0490 | possible feature_name: cytochrome oxidase subunit 2\n",
      "2025-06-06 18:29:22,298 - DEBUG - \tt: CDS | p: cytochrome oxidase subunit 2 | a: ID=cds-AFE02866.1;Parent=gene-AFUA_m0490;Dbxref=NCBI_GP:AFE02866.1;Name=AFE02866.1;gbkey=CDS;locus_tag=AFUA_m0490;product=cytochrome oxidase subunit 2;protein_id=AFE02866.1;transl_table=4\n",
      "2025-06-06 18:29:22,298 - DEBUG - \tlabel: MT-CO2, gene_dict[feature_name]: GeneDescription(label='MT-CO2', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,299 - DEBUG - \tgene-AFUA_m0490 | inserting from missing_dbxref_unk_problems, source: cytochrome oxidase subunit 2 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,300 - DEBUG -  gene-AFUA_m0500 | possible feature_name: tRNA-Arg\n",
      "2025-06-06 18:29:22,305 - DEBUG - \tt: tRNA | p: tRNA-Arg | a: ID=rna-AFUA_m0500;Parent=gene-AFUA_m0500;Note=anticodon ACG;gbkey=tRNA;locus_tag=AFUA_m0500;product=tRNA-Arg\n",
      "2025-06-06 18:29:22,306 - DEBUG - \tlabel: MT-TR, gene_dict[feature_name]: GeneDescription(label='MT-TR', c=None, source='MANUAL')\n",
      "2025-06-06 18:29:22,306 - DEBUG - \tgene-AFUA_m0500 | inserting from missing_dbxref_unk_problems, source: tRNA-Arg | type: tRNA\n",
      "\n",
      "2025-06-06 18:29:22,308 - DEBUG -  gene-AFUA_m0510 | possible feature_name: NADH dehydrogenase subunit 4L\n",
      "2025-06-06 18:29:22,310 - DEBUG - \tt: CDS | p: NADH dehydrogenase subunit 4L | a: ID=cds-AFE02867.1;Parent=gene-AFUA_m0510;Dbxref=NCBI_GP:AFE02867.1;Name=AFE02867.1;gbkey=CDS;locus_tag=AFUA_m0510;product=NADH dehydrogenase subunit 4L;protein_id=AFE02867.1;transl_table=4\n",
      "2025-06-06 18:29:22,310 - DEBUG - \tlabel: MT-ND4L, gene_dict[feature_name]: GeneDescription(label='MT-ND4L', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,311 - DEBUG - \tgene-AFUA_m0510 | inserting from missing_dbxref_unk_problems, source: NADH dehydrogenase subunit 4L | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,313 - DEBUG -  gene-AFUA_m0520 | possible feature_name: NADH dehydrogenase subunit 5\n",
      "2025-06-06 18:29:22,315 - DEBUG - \tt: CDS | p: NADH dehydrogenase subunit 5 | a: ID=cds-AFE02868.1;Parent=gene-AFUA_m0520;Dbxref=NCBI_GP:AFE02868.1;Name=AFE02868.1;gbkey=CDS;locus_tag=AFUA_m0520;product=NADH dehydrogenase subunit 5;protein_id=AFE02868.1;transl_table=4\n",
      "2025-06-06 18:29:22,315 - DEBUG - \tlabel: MT-ND5, gene_dict[feature_name]: GeneDescription(label='MT-ND5', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,316 - DEBUG - \tgene-AFUA_m0520 | inserting from missing_dbxref_unk_problems, source: NADH dehydrogenase subunit 5 | type: CDS\n",
      "\n",
      "2025-06-06 18:29:22,318 - DEBUG -  gene-AFUA_m0530 | possible feature_name: NADH dehydrogenase subunit 2\n",
      "2025-06-06 18:29:22,319 - DEBUG - \tt: CDS | p: NADH dehydrogenase subunit 2 | a: ID=cds-AFE02869.1;Parent=gene-AFUA_m0530;Dbxref=NCBI_GP:AFE02869.1;Name=AFE02869.1;gbkey=CDS;locus_tag=AFUA_m0530;product=NADH dehydrogenase subunit 2;protein_id=AFE02869.1;transl_table=4\n",
      "2025-06-06 18:29:22,319 - DEBUG - \tlabel: MT-ND2, gene_dict[feature_name]: GeneDescription(label='MT-ND2', c=None, source='NCBI')\n",
      "2025-06-06 18:29:22,320 - DEBUG - \tgene-AFUA_m0530 | inserting from missing_dbxref_unk_problems, source: NADH dehydrogenase subunit 2 | type: CDS\n",
      "\n"
     ]
    }
   ],
   "source": [
    "temp_unk = {}\n",
    "label_count = 0\n",
    "change_gene_dict = False\n",
    "log.debug(\n",
    "    \"missing_dbxref: matching probable 'child feature + parent gene' pair (2nd pass, on the an original gff3, using all the features)\"\n",
    ")\n",
    "for an in unk_dict.keys():\n",
    "    gene_ids = unk_dict[an]\n",
    "    log.debug(f\"AN: {an}| gene_ids: {gene_ids}\")\n",
    "    an_path = DATA_DIR / f\"{an}{gff_suffix}\"\n",
    "\n",
    "    df = gdt.gff3_utils.load_gff3(an_path, usecols=gdt.GFF3_COLUMNS)\n",
    "    df = gdt.gff3_utils.filter_orfs(df) if remove_orfs else df\n",
    "\n",
    "    df[\"gene_id\"] = df[\"attributes\"].str.extract(RE_ID, expand=False)\n",
    "    df[\"name\"] = df[\"attributes\"].str.extract(RE_NAME, expand=False)\n",
    "    df[\"product\"] = df[\"attributes\"].str.extract(RE_PRODUCT, expand=False)\n",
    "    df[\"description\"] = df[\"attributes\"].str.extract(RE_DESCRIPTION, expand=False)\n",
    "    df[\"gene\"] = df[\"attributes\"].str.extract(RE_GENE, expand=False)\n",
    "    df[\"parent\"] = df[\"attributes\"].str.extract(RE_PARENT, expand=False)\n",
    "    df[\"gene_synonym\"] = df[\"attributes\"].str.extract(RE_GENE_SYNONYM, expand=False)\n",
    "    df[\"note\"] = df[\"attributes\"].str.extract(RE_NOTE, expand=False)\n",
    "\n",
    "    df[\"feature_name\"] = (\n",
    "        df[\"gene\"]\n",
    "        .fillna(df[\"product\"])\n",
    "        .fillna(df[\"description\"])\n",
    "        .fillna(df[\"name\"])\n",
    "        .fillna(df[\"note\"])\n",
    "        .fillna(df[\"gene_synonym\"])\n",
    "    )\n",
    "\n",
    "    for gene_id in gene_ids:\n",
    "        candidates = df[df[\"parent\"] == gene_id]\n",
    "        if len(candidates) >= 1:\n",
    "            if candidates[\"feature_name\"].nunique() > 1:\n",
    "                log.debug(\n",
    "                    f\" {gene_id} | more than one feature with parent {gene_id} found, but with feature_name conflict\"\n",
    "                )\n",
    "                log.debug(\n",
    "                    f\" {gene_id} | feature_names: {candidates['feature_name'].unique()}\"\n",
    "                )\n",
    "                log.debug(\n",
    "                    f\" {gene_id} | choosing the first one: {candidates['feature_name'].iloc[0]}\"\n",
    "                )\n",
    "\n",
    "            feature_name = candidates[\"feature_name\"].iloc[0]\n",
    "            log.debug(f\" {gene_id} | possible feature_name: {feature_name}\")\n",
    "            [\n",
    "                log.debug(f\"\\tt: {x.type} | p: {x.feature_name} | a: {x.attributes}\")\n",
    "                for x in candidates.itertuples()\n",
    "            ]\n",
    "\n",
    "            if feature_name in gene_dict:\n",
    "                change_gene_dict = True\n",
    "                label = gene_dict[feature_name].label\n",
    "                log.debug(\n",
    "                    f\"\\tlabel: {label}, gene_dict[feature_name]: {gene_dict[feature_name]}\"\n",
    "                )\n",
    "                log.debug(\n",
    "                    f\"\\t{gene_id} | inserting from missing_dbxref_unk_problems, source: {feature_name} | type: {candidates['type'].iloc[0]}\\n\"\n",
    "                )\n",
    "                gene_dict[gene_id] = gdt.gene_dict_impl.GeneGeneric(\n",
    "                    label=label,\n",
    "                    an_sources=[an],\n",
    "                    c=f\"insertion from missing_dbxref_unk_problems, source: {feature_name} | type: {candidates['type'].iloc[0]}\",\n",
    "                )\n",
    "\n",
    "            else:\n",
    "                log.debug(f\"\\tproduct not in gene_dict: {feature_name}\\n\")\n",
    "                log.debug(\"\\tchecking in temp_unk\")\n",
    "\n",
    "                if feature_name in temp_unk:\n",
    "                    log.debug(\n",
    "                        f\"\\t{gene_id} | found in temp_unk, inserting gene_id in there\"\n",
    "                    )\n",
    "                    label = temp_unk[feature_name].label\n",
    "                    gene_dict[gene_id] = gdt.gene_dict_impl.GeneGeneric(\n",
    "                        label=label,\n",
    "                        an_sources=[an],\n",
    "                        c=f\"insertion from missing_dbxref_unk_problems, source: {feature_name} | type: {candidates['type'].iloc[0]}\",\n",
    "                    )\n",
    "                else:\n",
    "                    log.debug(\n",
    "                        f\"\\t{gene_id} | not found in temp_unk, creating new entry\"\n",
    "                    )\n",
    "                    label_count += 1\n",
    "                    label = f\"{organelle_type}-TEMP-{label_count}\"\n",
    "                    temp_unk[feature_name] = gdt.gene_dict_impl.GeneDescription(\n",
    "                        label=label, source=\"MANUAL\", c=None\n",
    "                    )\n",
    "\n",
    "                    temp_unk[gene_id] = gdt.gene_dict_impl.GeneGeneric(\n",
    "                        label=label,\n",
    "                        an_sources=[an],\n",
    "                        c=f\"insertion from missing_dbxref_unk_problems, source: {feature_name} | type: {candidates['type'].iloc[0]}\",\n",
    "                    )\n",
    "        else:\n",
    "            log.debug(f\" {gene_id} | no features with parent {gene_id} found\")\n",
    "            log.debug(\" adding it to UNKNOWN label \")\n",
    "            temp_unk[gene_id] = gdt.gene_dict_impl.GeneGeneric(\n",
    "                label=f\"{organelle_type}-UNKNOWN\",\n",
    "                an_sources=[an],\n",
    "                c=f\"unknown gene_id from {an}{gff_suffix} | \"\n",
    "                f\"a: {df[df['gene_id'] == gene_id]['attributes'].iloc[0] if not df[df['gene_id'] == gene_id].empty else 'N/A'}\",\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a790c381",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-06-06 18:29:47,285 - INFO - Writing TEMP Mapping GDT file: /home/brenodupin/matheus/gdt/sandbox/fungi_mt_model/misc/gdt/TEMP_Mapping_01.gdt | Iteration: 1\n"
     ]
    }
   ],
   "source": [
    "if temp_unk:\n",
    "    temp_path = get_most_recent_gdt(GDT_dir, prefix=\"TEMP_Mapping_\")\n",
    "    new_path, map_iteration = increment_gdt_file(temp_path)\n",
    "    log.info(\n",
    "        f\"Writing TEMP Mapping GDT file: {new_path} | Iteration: {map_iteration}\"\n",
    "    )\n",
    "    temp_unk[\"gdt_info\"] = gdt.gene_dict_impl.get_gene_dict_info(temp_unk)\n",
    "    temp_unk[\"gdt_header\"] = [\n",
    "        \"version 0.0.2\",\n",
    "        f\"TEMP Mapping - {map_iteration}\",\n",
    "        \"Automagically generated by AN_missing_dbxref.ipynb | TEMP Mapping child features to parent genes\",\n",
    "    ]\n",
    "    gdt.gene_dict_impl.write_gdt_file(temp_unk, new_path, overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74b38296",
   "metadata": {},
   "outputs": [],
   "source": [
    "if change_gene_dict:\n",
    "    log.debug(\"gene_dict changed, incrementing gdt file and writing it\")\n",
    "    most_recent_gdt_file, _ = increment_gdt_file(most_recent_gdt_file)\n",
    "    log.info(f\"Writing GDT file: {most_recent_gdt_file}\")\n",
    "    gene_dict[\"gdt_info\"] = gdt.gene_dict_impl.get_gene_dict_info(gene_dict)\n",
    "    gene_dict[\"gdt_header\"].append(\n",
    "        f\"{datetime.now().strftime('%Y-%m-%d %H:%M')} - Data added from TEMP Mapping\"\n",
    "    )\n",
    "    gdt.gene_dict_impl.write_gdt_file(gene_dict, most_recent_gdt_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "857707e0",
   "metadata": {},
   "source": [
    "### Genes Discard using gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "9ba79d2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_string = \"discard-\"\n",
    "genes_to_remove = \"genome_features_to_remove_2.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f1e57d4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_gene_ids = {}\n",
    "with open(MISC_DIR / genes_to_remove, \"r\") as f:\n",
    "    for line in f:\n",
    "        if (\n",
    "            not line.strip()\n",
    "            or line.startswith(\"#\")\n",
    "            or line.startswith(\"[\")\n",
    "            or \"#gd\" in line\n",
    "        ):\n",
    "            continue\n",
    "\n",
    "        gene_id, an = line.split(\"#c\", 1)[0].split(\"#gn\", 1)\n",
    "        gene_id = gene_id.strip()\n",
    "        an = an.strip()\n",
    "        if not an or not gene_id:\n",
    "            raise ValueError(f\"Error: {line} - wtf (TODO, remove)?\")\n",
    "\n",
    "        if an not in remove_gene_ids:\n",
    "            remove_gene_ids[an] = set([gene_id])\n",
    "        else:\n",
    "            remove_gene_ids[an].add(gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "f1d59f95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'JQ346808.1': {'gene-AFUA_m0390'}}"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "remove_gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "43d1705c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JQ346808.1 Done!\n"
     ]
    }
   ],
   "source": [
    "for an in remove_gene_ids.keys():\n",
    "    an_path = DATA_DIR / f\"{an}{gff_suffix}\"\n",
    "    with open(an_path, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    headers, index = [], 0\n",
    "    while lines[index].startswith(\"#\"):\n",
    "        headers.append(lines[index].strip())\n",
    "        index += 1\n",
    "\n",
    "    pattern = re.compile(\"|\".join([f\"ID={x};\" for x in remove_gene_ids[an]]))\n",
    "    contents = []\n",
    "\n",
    "    for line in lines[index:]:\n",
    "        if not (line := line.strip()):\n",
    "            continue\n",
    "        line = line.split(\"\\t\")\n",
    "\n",
    "        # line[2] is type line, line[8] is attributes\n",
    "        if pattern.search(line[8]):\n",
    "            if remove_string not in line[2]:\n",
    "                line[2] = remove_string + line[2]\n",
    "\n",
    "        contents.append(\"\\t\".join(line))\n",
    "\n",
    "    with open(an_path, \"w\") as f:\n",
    "        f.write(\"\\n\".join(headers))\n",
    "        f.write(\"\\n\")\n",
    "        f.write(\"\\n\".join(contents))\n",
    "        f.write(\"\\n\\n\")\n",
    "\n",
    "    print(f\"{an} Done!\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
